---
title: "01 FirstAnalysis"
author: "Nick Diercksen"
output:
  html_notebook:
    code_folding: hide
    toc: yes
    toc_float: yes
    df_print: paged
editor_options: 
  chunk_output_type: inline
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(dplyr)
library(ggplot2)
library(treemap)
library(DESeq2)
library(ggfortify) # autoplot (PCA)
library(ggpubr)

# Heatmaps
# ## Github & ComplexHeatmap ----
if (! "ComplexHeatmap" %in% row.names(installed.packages())) {
  devtools::install_github("jokergoo/ComplexHeatmap", force = TRUE)
  # BiocManager::install("ComplexHeatmap")
}
library(ComplexHeatmap)
library(circlize)
library(RColorBrewer)
library(gplots) # (heatmap.2)
library(SummarizedExperiment)
```

# Overview

## Preparing data

```{r}
gastroc <- readr::read_tsv("./data/counts/readcount_genename_gastroc.xls", show_col_types = F) %>%
  tibble::column_to_rownames(var = "gene_id")
counts.gastroc <- gastroc[, 1:12]

# TODO: fix annotations: several entries are incomplete whereas the gastroc seems to be complete
soleus <- readr::read_tsv("./data/counts/readcount_genename_soleus.xls", show_col_types = F) %>%
  tibble::column_to_rownames(var = "gene_id")
counts.soleus <- soleus[, 1:12]

```

## Treemap

Both counts have the same genes analysed.
These amount to nearly 54,000 genes, which corresponds to the whole genome of mice. 

```{r}
gene_biotypes <- table(gastroc$gene_biotype) %>%
  as.data.frame() %>%
  arrange(desc(Freq))

treemap::treemap(gene_biotypes,
                 index = "Var1",
                 vSize = "Freq",
                 title = "Gene biotypes")
```

# Filtering

## Overview

For a broad overview all genes with `rowSums() < 10` will be displayed. This is usually similar to omitting genes with zero counts of more than 30%

```{r}
rs.gastroc <- rowSums(counts.gastroc) %>%
  as.data.frame() %>%
  dplyr::rename("rowSums" = ".")
n.omit.gastroc <- sum(rs.gastroc < 10)
title <- paste0("rowSums low expressed genes, gastroc \n",
                n.omit.gastroc, " genes will be omitted (~",
                floor(n.omit.gastroc / nrow(rs.gastroc) * 100), " %)")

p.gastroc <- filter(rs.gastroc, rowSums < 20) %>%
  ggplot(., aes(x=rowSums))+
  geom_histogram(color="black", fill="white")+
  geom_vline(aes(xintercept=10, color="red"),
             linetype="dashed") +
  ggtitle(title) + 
  theme(legend.position = "none")

# ---
rs.soleus <- rowSums(counts.soleus) %>%
  as.data.frame() %>%
  dplyr::rename("rowSums" = ".")
n.omit.soleus <- sum(rs.soleus < 10)
title <- paste0("rowSums low expressed genes, soleus \n",
                n.omit.soleus, " genes will be omitted (~",
                floor(n.omit.gastroc / nrow(rs.gastroc) * 100), " %)")

p.soleus <- filter(rs.soleus, rowSums < 20) %>%
  ggplot(., aes(x=rowSums))+
  geom_histogram(color="black", fill="white")+
  geom_vline(aes(xintercept=10, color="red"),
             linetype="dashed") +
  ggtitle(title)+ 
  theme(legend.position = "none")

p.gastroc
p.soleus

```

## omitting genes (actual filtering)

omitting genes with zero counts in more than 30% of the samples

```{r echo=FALSE}
omitZeroCounts <- function(counts.raw, share=0.3) {
  keep <- (counts.raw != 0) %>%
    rowSums() > round(share * ncol(counts.raw))
  return(counts.raw[keep,])
}

counts.gastroc <- omitZeroCounts(counts.gastroc)
counts.soleus <- omitZeroCounts(counts.soleus)
```



# Quality assessment

## Library sizes

-   too many zero counts
-   different total counts

```{r}
plot.library_size <- function(counts, title="") {
  cs <- colSums(counts) %>%
    as.data.frame() %>%
    dplyr::rename("colsums" = ".") %>%
    tibble::rownames_to_column(var = "sample")
  
  ggplot(cs, aes(x = sample, y = colsums)) +
    geom_bar(stat = "identity") +
    theme(axis.text.x=element_text(color = "black", angle=45, vjust=.8, hjust=0.8)) + 
    ggtitle(paste0("total counts, ", title))
}


plt.libsize.gastroc <- plot.library_size(counts.gastroc, title = "gastroc") +
  ylim(0, 4e07)
plt.libsize.soleus <- plot.library_size(counts.soleus, title = "soleus") +
  ylim(0, 4e07)

ggpubr::ggarrange(plt.libsize.gastroc, plt.libsize.soleus,
                    ncol = 1, nrow = 2)
```

not sure what to make of this, but its there.


### count distribution boxplots

Preparing the sample by applying log2 and loading metadata files

```{r echo=FALSE, message=FALSE}
logcounts.gastroc <- log2(counts.gastroc + 1) # avoiding -Inf with ++
logcounts.soleus <- log2(counts.soleus + 1) # avoiding -Inf with ++

metadata.gastroc <- readr::read_csv2("./data/counts/sample_table_gastroc.csv") %>%
  tibble::column_to_rownames("run")
metadata.soleus <- readr::read_csv2("./data/counts/sample_table_soleus.csv") %>%
  tibble::column_to_rownames("run")
```


Resulting Plots:

```{r message=FALSE}
plotBoxLog <- function(logcounts, metadata) {
  stack(logcounts) %>%
    merge(metadata, by.x = "ind", by.y = "sample_name") %>% 
    ggplot( aes(x=ind, y=values, color=genotype)) +
    geom_boxplot() + 
    ylab("Log2(Counts)") +
    xlab(element_blank()) +
    theme(axis.text.x=element_text(angle=45, vjust=.8, hjust=0.8)) +
    geom_hline(yintercept = median(as.matrix(logcounts.gastroc)), col="blue" )
}

p1 <- plotBoxLog(logcounts.gastroc, metadata.gastroc) + 
  ggtitle("gastroc")
p2 <- plotBoxLog(logcounts.soleus, metadata.soleus) + 
  ggtitle("soleus") +
  theme(axis.ticks.y = element_blank(),
        axis.text.y = element_blank()) +
  rremove("ylab")
```


```{r echo=FALSE, message=FALSE}
pdf(NULL) # avoiding blank plot created by ggarrange
p <- ggpubr::ggarrange(p1, p2, ncol=2, common.legend = TRUE, legend = "right")
x = dev.off()  # see above
p
```



### using DESeq2::rlog to normalize

Just for exploration purpose, but the resulting data is not used in the
following analysis.

```{r message=FALSE}
p1 <- DESeq2::rlog(as.matrix(counts.gastroc)) %>%
  as.data.frame() %>%
  plotBoxLog(metadata.gastroc) + 
  ggtitle("gastroc") +
  ylab("rlog(counts)")

p2 <- DESeq2::rlog(as.matrix(counts.soleus)) %>%
  as.data.frame() %>%
  plotBoxLog(metadata.soleus) + 
  ggtitle("soleus") +
  theme(axis.ticks.y = element_blank(),
        axis.text.y = element_blank()) +
  rremove("ylab")
```


```{r echo=FALSE, message=FALSE}
pdf(NULL) # avoiding blank plot created by ggarrange
p <- ggpubr::ggarrange(p1, p2, ncol=2, common.legend = TRUE, legend = "right")
x = dev.off()  # see above
p
```

# most variable genes (top 1000)
To obtain the most variable genes, a z-score will be applied to then take the
sd and filter for the top 1000

$\frac{x - mean(x)}{sd(x)}$

```{r}
mostVariableRows <- function(M, ntop=1000) {
  s <- apply( M,1,sd )      # standard deviation
  µ <- apply( M, 1, mean )  # mean
  M.z <- (M - µ) / s         # zscore
  
  # sd ordered descending
  sdev <- apply(counts.gastroc, 1, sd)
  M.top <- M[order(sdev, decreasing = T)[1:ntop] , ]
  return(M.top)
}

counts.gastroc.top <- mostVariableRows(counts.gastroc)
counts.soleus.top <- mostVariableRows(counts.soleus)
```


## visualization with Heatmaps
To visualize the top 1000 most variable genes, I tried to generate some Heatmaps. (This time not using DESeq2::rlog, but just `log2(counts)`. Using the raw counts
I could try to just apply `log2` to all counts and then create a heatmap.

### `ComplexHeatmap`
The following Heatmaps are done with `ComplexHeatmap` but are very bad looking.

#### gastroc

```{r message=FALSE}
# Get some nicer colours
mypalette <- brewer.pal(11, "RdYlBu")
# http://colorbrewer2.org/#type=sequential&scheme=BuGn&n=3
morecols <- colorRampPalette(mypalette)


chm.gastroc <- Heatmap(
    log2(counts.gastroc.top + 1) %>% as.matrix(),
    row_title = "genes",
    name = "log2",
    show_row_names = FALSE,
    column_title = "samples",
    col = rev(morecols(50)),
    column_title_side = "bottom"
    # row_dend_side = "right",
  )

# saving the CHM as png, to adjust the resolution (I don't know another method)
png(
  "./plots/01_CHM_top_variable_genes.gastroc.png",
  width = 10,
  height = 7.5,
  res = 150,
  units = "in"
)
draw(chm.gastroc)
invisible(dev.off())
```

![](./plots/01_CHM_top_variable_genes.gastroc.png)

#### soleus

```{r message=FALSE}
chm.soleus <- Heatmap(
    log2(counts.soleus.top + 1) %>% as.matrix(),
    row_title = "genes",
    name = "log2",
    show_row_names = FALSE,
    column_title = "samples",
    col = rev(morecols(50)),
    column_title_side = "bottom"
    # row_dend_side = "right",
  )

# saving the CHM as png, to adjust the resolution (I don't know another method)
png(
  "./plots/01_CHM_top_variable_genes.soleus.png",
  width = 10,
  height = 7.5,
  res = 150,
  units = "in"
)
draw(chm.soleus)
invisible(dev.off())
```

![](./plots/01_CHM_top_variable_genes.soleus.png)



### `gplots::heatmap.2`

With this package, the heatmaps are seemingly generated differently, even though
I thought both packages use `hclust()`. Not sure yet what is the difference.

```{r}
# Set up colour vector for celltype variable
col.cell <- rep(c("#00C3C6"), 12)
col.cell[metadata.gastroc$genotype == "KO"] = "#FF6C67"

heatmap.2(log2(as.matrix(counts.gastroc.top + 1)),
          labRow = FALSE,
          col=rev(morecols(50)),
          ColSideColors=col.cell, scale="row",
          trace="column",
          main="gastroc",
          lmat=rbind(c(5,4), c(3,2), c(0,1)),
          lhei=c(2.2,4,0.2))

heatmap.2(log2(as.matrix(counts.soleus.top + 1)), 
          labRow = FALSE,
          col=rev(morecols(50)),
          ColSideColors=col.cell, scale="row",
          trace="column",
          main="soleus",
          lmat=rbind(c(5,4), c(3,2), c(0,1)),
          lhei=c(2.2,4,0.2))
```




# PCA

## gastroc

the pca is performed on the raw data (not z-scored)
=> Thus using the `scale.=T` argument?

```{r message=FALSE}
pca <- prcomp(t(counts.gastroc.top), scale. = T)

pca.data <- data.frame(Sample = rownames(pca$x),
                       X = pca$x[, 1],
                       Y = pca$x[, 2]) %>%
  mutate(type = substr(Sample, 1,2))

plt <-
  autoplot(
    pca,
    data = pca.data,
    colour = 'type',
    label.show.legend = FALSE
  ) +
  ggtitle("PCA gastroc")
# ggsave(filename = "./plots/04_pca_gastroc_filtered.png", plt, dpi=300)
plt
```

## soleus

```{r message=FALSE}
pca <- prcomp(t(counts.soleus.top), scale. = T)

pca.data <- data.frame(Sample = rownames(pca$x),
                       X = pca$x[, 1],
                       Y = pca$x[, 2]) %>%
  mutate(type = substr(Sample, 1,2))

plt <-
  autoplot(
    pca,
    data = pca.data,
    colour = 'type',
    shape = 17,
    label.show.legend = FALSE
  ) +
  ggtitle("PCA soleus")
# ggsave(filename = "./plots/04_pca_soleus_filtered.png", plt, dpi=300)
plt
```

## both tissues
Here also taking the top 1000 most variable genes of each tissue and merging them.

```{r message=FALSE}
# combine both count tables
counts.both <-
  merge(counts.gastroc.top,
        counts.soleus.top,
        by = 0,
        suffixes = c("_gastroc", "_soleus")) %>%
  tibble::column_to_rownames("Row.names")

pca <- prcomp(t(counts.both), scale. = T)

pca.data <- data.frame(Sample = rownames(pca$x),
                       X = pca$x[, 1],
                       Y = pca$x[, 2]) %>%
  mutate(type = substr(Sample, 1, 2),
         tissue = stringr::str_extract(Sample,"[:alpha:]+$"))

plt <-
  autoplot(
    pca,
    data = pca.data,
    colour = 'type',
    shape = 'tissue',
    label.show.legend = FALSE
  ) +
  ggtitle("PCA both tissues")
# ggsave(filename = "./plots/04_pca_both_tissues_filtered.png", plt, dpi=300)
plt
```


# cleaning up

## SE
```{r SummarizedExperiment}
# ordering data.frames by rowname
gastroc <- gastroc[order(rownames(gastroc)),]
soleus <- soleus[order(rownames(soleus)),]

# rowData
rowdata.gastroc <- gastroc[, 13:21]
rowdata.soleus <- rowdata.gastroc # should be the same I think
rowdata.gastroc$filtered = rownames(rowdata.gastroc) %in% rownames(counts.gastroc)
rowdata.soleus$filtered  = rownames(rowdata.soleus) %in% rownames(counts.soleus)

se.gastroc <- SummarizedExperiment(
  assays = list(
    "gastroc" = as.matrix(gastroc[, 1:12])
  ),
  rowData = rowdata.gastroc,
  colData = metadata.gastroc[, 3, drop = F] # only genotype is available as information
)

se.soleus <- SummarizedExperiment(
  assays = list(
    "soleus" = as.matrix(soleus[, 1:12])
  ),
  rowData = rowdata.soleus,
  colData = metadata.soleus[, 3, drop = F] # only genotype is available as information
)
```

the filtered assays can be accessed as following:
```{r SE-filter, eval=FALSE}
counts.preFiltered <- se[rowData(se)$filtered,] %>% assay()
```

## save R Objects

```{r}
saveRDS(se.gastroc,             file = "./data/Robjects/01_se.gastroc.rds")
saveRDS(se.soleus,             file = "./data/Robjects/01_se.soleus.rds")

# (redundant)
saveRDS(counts.gastroc, file = "./data/Robjects/counts.gastroc.rds")
saveRDS(counts.soleus,  file = "./data/Robjects/counts.soleus.rds")
```



# Questions
Most variable genes:

* which heatmap function should I use and what should I change about the plots?
* should I include the traces?
* should I completely omit the heatmaps (visualisation of most variable genes?)

PCA:

* I was not sure anymore if the PCA should be performed with normalized or not normalized counts? (Last week I thought we said not normalized, but today I think you said normalized?) Just want to double check, sorry.
* if the data is not normalized, does it mean to then use the `scale=T` argument?