---
title: "01_FirstAnalysis"
author: "Nick Diercksen"
date: "Last compiled on `r format(Sys.time())`"
output:
  html_notebook:
    code_folding: hide
    toc: yes
    toc_float: yes
    df_print: paged
  html_document:
    toc: yes
    df_print: paged
editor_options: 
  chunk_output_type: inline
params:
  condition_pal: !r c("WT"= "#00C3C6","KO"= "#FF6C67")
  tissue_pal: !r c(gastroc = "orange", soleus="purple")
  heatmap_pal: !r rev(RColorBrewer::brewer.pal(11, "RdYlBu"))
  treemap_pal: !r colorRampPalette(RColorBrewer::brewer.pal(11, "Paired"))
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, results = "hold")
library(dplyr)
library(eulerr)
library(ggplot2)
library(ggpubr)
library(RColorBrewer)
library(treemap)

if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
if (!require("DESeq2", quietly = TRUE))
  BiocManager::install("DESeq2")
if (!require("SummarizedExperiment", quietly = TRUE))
  BiocManager::install("SummarizedExperiment")

library(DESeq2)
library(SummarizedExperiment)
```


```{r plottingSettings, include=FALSE}
# for custom CHM and PCA plotting functions
source(file.path("code", "visualization_functions.R"))

ggplot2::theme_set(theme_bw())

plot.dir <- file.path("plots", "01 Analysis and Filtering")
```

# Overview

## Preparing data

### reading in

```{r}
gastroc <-
  readr::read_tsv("./data/counts/readcount_genename_gastroc.xls",
                  show_col_types = F) %>%
  arrange(gene_id) %>% 
  tibble::column_to_rownames(var = "gene_id")
counts.gastroc <- gastroc[, 1:12]

# TODO: fix annotations: several entries are incomplete whereas the gastroc seems to be complete
soleus <-
  readr::read_tsv("./data/counts/readcount_genename_soleus.xls",
                  show_col_types = F) %>%
  arrange(gene_id) %>% 
  tibble::column_to_rownames(var = "gene_id")
counts.soleus <- soleus[, 1:12]

```


### comparing annotations between the files of the respective tissues

```{r compareAnnotations, results = 'markup'}
# soleus contains `NA` values, whereas gastroc does not:
anyNA(gastroc) # FALSE
anyNA(soleus)  # TRUE

# rownames (gene_id)
assertthat::are_equal(rownames(gastroc), rownames(soleus)) # TRUE


# COMPARE COMLUMNS:
col_idx <- c(1,10,2,11,3,12,4,13,5,14,6,15,7,16,8,17,9,18)
compare.df <- merge(gastroc[13:21], soleus[13:21], by = 0, suffixes = c(".ga", ".so")) %>% 
  tibble::column_to_rownames("Row.names") %>% 
  .[,col_idx]


# `gene_name`
assertthat::are_equal(gastroc$gene_name, soleus$gene_name) # FALSE

# `gene_start`
assertthat::are_equal(compare.df$gene_start.ga, compare.df$gene_start.so) # FALSE
# filter(compare.df, gene_start.ga != gene_start.so) %>% View() # 504 entries

# `gene_end`
assertthat::are_equal(compare.df$gene_end.ga, compare.df$gene_end.so) # FALSE
# filter(compare.df, gene_end.ga != gene_end.so) %>% View() # 506 entries

# `gene_chr`
assertthat::are_equal(compare.df$gene_chr.ga, compare.df$gene_chr.so) # FALSE
# filter(compare.df, gene_chr.ga != gene_chr.so) %>% View() # 58 entries (in `so` no `chr` available)

# `gene_strand`
assertthat::are_equal(compare.df$gene_strand.ga, compare.df$gene_strand.so) # FALSE
# filter(compare.df, gene_strand.ga != gene_strand.so) %>% View() # 33 entries (all inverse, where 32 are "+" in ga)

# `gene_biotype`
assertthat::are_equal(compare.df$gene_biotype.ga, compare.df$gene_biotype.so) # FALSE
# filter(compare.df, gene_biotype.ga != gene_biotype.so) %>% View() # 155 entries

# `gene_description`
assertthat::are_equal(compare.df$gene_description.ga, compare.df$gene_description.so) # FALSE
# filter(compare.df, gene_description.ga == gene_description.so) %>% View() # 58 entries have no description at all, gastroc has no description at all

# `tf_family`
assertthat::are_equal(compare.df$tf_family.ga, compare.df$tf_family.so) # FALSE
# filter(compare.df, tf_family.ga != tf_family.so) %>% View() # 11096 entries, 

rm(compare.df)
```

Genes for both tissues have the same ensembl id, and have in general the same annotations.
But each annotation (column) has entries which are inconsistent!

The focus in the subsequent analysis will remain using the gene_ids while most annotations, except the occasional gene name for plotting will not be used.


# Filtering

## Overview

For a broad overview of the the gene expression distribution, each expression value (of the samples) will be summed up for each gene (combining all samples).
For comparison to other analysis where often a threshold of `10` is chosen to omit low expressed genes:

```{r densityRowsumsOverlayed}
# pair rowsums with tissue for ggplot
rs.df <- data.frame(rowSums = c(rowSums(counts.gastroc), rowSums(counts.soleus)),
                    tissue = rep(c("gastrocnemius", "soleus"),
                                 times = c(
                                   nrow(counts.gastroc), nrow(counts.gastroc)
                                 )))

cutoff <- 10

p <- ggplot(rs.df, aes(x=log(rowSums + 1), fill=tissue, color=tissue, linetype=tissue)) + 
  geom_density(alpha=0.3) +
  scale_fill_manual(values = unname(params$tissue_pal), aesthetics = c("fill", "color")) +
  geom_vline(xintercept = log(cutoff +1), linetype = "solid", color="red", alpha=0.5) + 
  annotate("text", x=log(cutoff+1), y = 0.5, label="rowSums=10", color="red") +
  ylab("density")
p
```


```{r densityRowsumsOverlayed_save, include=FALSE}
ggsave(file.path(plot.dir, "01_rowsums_combined.svg"), p, width=16, height=8, units = "cm")
ggsave(file.path(plot.dir, "01_rowsums_combined.png"), p, width=16, height=8, units = "cm")
```


Determining the actual number of genes, that would remain after applying that cutoff:
```{r rowsums_cutoff_amount, results='hold', collapse=FALSE}
filter(rs.df, rowSums > 10, tissue == "gastrocnemius") %>% nrow()
filter(rs.df, rowSums > 10, tissue == "soleus") %>% nrow()
```



## omitting genes (actual filtering)

We chose to omit genes having zero counts in more than 30% of the samples.
This is a similar approach than above, but cannot be visualized as nicely.

The zero-counts were detected and omitted for each tissue separately, which meant
that the total count of genes kept differ slightly.

In a future attempt of adjusting this analysis, the union of those genes could be taken,
so that the same genes were to be kept in both tissues.


```{r omit_zero_counts}
omitZeroCounts <- function(counts.raw, share=0.3) {
  keep <- (counts.raw != 0) %>%
    rowSums() > round(share * ncol(counts.raw))
  return(counts.raw[keep,])
}

counts.gastroc <- omitZeroCounts(counts.gastroc)
counts.soleus <- omitZeroCounts(counts.soleus)
```


## Venn (Euler) Diagram

```{r vennFiltered}
sets <- list(
  "all" = rownames(gastroc),
  "gastrocnemius" = rownames(counts.gastroc),
  "soleus" = rownames(counts.soleus)
)

venn.colors <- c(genes = "white", params$tissue_pal)

p <- plot(
  eulerr::euler(sets),
  fills = venn.colors,
  quantities = T
)
p
```


```{r vennFiltered_save, include=FALSE}
ggsave(file.path(plot.dir, "01_filtered_genes.svg"), p)
```



## Treemap

The same genes were analysed for both tissues.
These amount to nearly 54,000 genes, which corresponds to the whole genome of mice. 

To see what genes were generally found expressed in both tissues the distribution of
gene biotypes is depicted in the following heatmap.

```{r gene_biotypes}
gene_biotypes <- table(gastroc$gene_biotype, dnn = c("biotype")) %>%
  as.data.frame() %>%
  mutate(
    biotype = stringr::str_replace_all(biotype, "_", " "),
    colors = params$treemap_pal(n()),
    ) %>% 
  arrange(desc(Freq))

# add percentage
gene_biotypes <- gene_biotypes %>% 
  mutate(
    share = round(Freq / sum(gene_biotypes$Freq),2) * 100,
    label = paste0(biotype, "\n", share, "%")
  )

treemap::treemap(gene_biotypes,
                 type = "color",
                 index = "label",
                 vSize = "Freq",
                 vColor = "colors",
                 title = "Gene biotypes",
                 fontsize.labels = 20)
# note: it occured sometimes, that when knitting the labels are omitted in the treemap
```


vs the biotypes of the remaining genes (after omitting the zero-count genes) which
occur in both tissues.


```{r gene_biotypes_filtered}
gene_ids <- unique(c(rownames(counts.gastroc), rownames(counts.soleus)))
gene_biotypes_filtered <- filter(gastroc, rownames(gastroc) %in% rownames(counts.gastroc)) %>% 
  select(gene_biotype) %>% 
  table(dnn = c("biotype")) %>%
  as.data.frame() %>%
  arrange(desc(Freq)) %>% 
  mutate(biotype = stringr::str_replace_all(biotype, "_", " ")) %>% 
  merge(gene_biotypes[, c("biotype", "colors")], by = "biotype") # obtain the same colors

# add percentage
gene_biotypes_filtered <- gene_biotypes_filtered %>% 
  mutate(
    share = round(Freq / sum(gene_biotypes_filtered$Freq),2) * 100,
    label = paste0(biotype, "\n", share, "%")
  )

treemap::treemap(gene_biotypes_filtered,
                 type = "color",
                 index = "label",
                 vColor = "colors",
                 vSize = "Freq",
                 title = "Gene biotypes, filtered",
                 fontsize.labels = 20)
```

The precise statistics of the changed distribution of biotypes can be seen in
the following table:

```{r comparing_biotypes}
gene_biotypes_cmp <-
  merge(
    gene_biotypes[, c(1, 2)],
    gene_biotypes_filtered[, c(1, 2)],
    by = "biotype",
    all = TRUE,
    suffixes = c(".all", ".filt")
  ) %>%
  mutate(
    "share"      = 100 * Freq.filt / Freq.all,
    "share.all"  = 100 * Freq.all / sum(gene_biotypes$Freq),
    "share.filt" = 100 * Freq.filt / sum(gene_biotypes_filtered$Freq)
  )
knitr::kable(gene_biotypes_cmp)
```



# Quality assessment

In this section the library sizes of the samples are compared with bar- and boxplots.

## Library sizes

-   too many zero counts
-   different total counts

```{r lib_sizes}
plot.library_size <- function(counts, title="") {
  cs <- colSums(counts) %>%
    as.data.frame() %>%
    dplyr::rename("colsums" = ".") %>%
    tibble::rownames_to_column(var = "sample")
  
  ggplot(cs, aes(x = sample, y = colsums)) +
    geom_bar(stat = "identity") +
    theme(axis.text.x=element_text(color = "black", angle=45, vjust=.8, hjust=0.8)) + 
    ggtitle(paste0("total counts, ", title))
}


plt.libsize.gastroc <- plot.library_size(counts.gastroc, title = "gastrocnemius") +
  ylim(0, 4e07)
plt.libsize.soleus <- plot.library_size(counts.soleus, title = "soleus") +
  ylim(0, 4e07)

ggpubr::ggarrange(plt.libsize.gastroc, plt.libsize.soleus,
                    ncol = 1, nrow = 2)
```



### count distribution boxplots

Preparing the sample by applying log2 and loading metadata files

```{r counts.distr.prep, message=FALSE}
logcounts.gastroc <- log2(counts.gastroc + 1) # avoiding -Inf with ++
logcounts.soleus <- log2(counts.soleus + 1) # avoiding -Inf with ++

metadata.gastroc <- readr::read_csv2("./data/counts/sample_table_gastroc.csv") %>%
  tibble::column_to_rownames("run")
metadata.soleus <- readr::read_csv2("./data/counts/sample_table_soleus.csv") %>%
  tibble::column_to_rownames("run")
```


Resulting Plots:

```{r counts.distr.plot, message=FALSE}
plotBoxLog <- function(logcounts, metadata) {
  stack(logcounts) %>%
    merge(metadata, by.x = "ind", by.y = "sample_name") %>% 
    ggplot( aes(x=ind, y=values, color=genotype)) +
    scale_color_manual(values = params$condition_pal) +
    geom_boxplot() + 
    ylab("Log2(Counts+1)") +
    xlab(element_blank()) +
    theme(axis.text.x=element_text(angle=45, vjust=.8, hjust=0.8)) +
    geom_hline(yintercept = median(as.matrix(logcounts)), col="blue" )
}

p1 <- plotBoxLog(logcounts.gastroc, metadata.gastroc) + 
  ggtitle("gastrocnemius")
p2 <- plotBoxLog(logcounts.soleus, metadata.soleus) + 
  ggtitle("soleus") +
  theme(axis.ticks.y = element_blank(),
        axis.text.y = element_blank()) +
  rremove("ylab")
```


```{r counts.distr.pdf, echo=FALSE, message=FALSE}
pdf(NULL) # avoiding blank plot created by ggarrange
p <- ggpubr::ggarrange(p1, p2, ncol=2, common.legend = TRUE, legend = "right")
x = dev.off()  # see above
p
```



### using DESeq2::rlog to normalize

Just for exploration purpose, but the resulting data is not used in the
following analysis.

```{r counts.rlog.plot, message=FALSE}
p1 <- DESeq2::rlog(as.matrix(counts.gastroc)) %>%
  as.data.frame() %>%
  plotBoxLog(metadata.gastroc) + 
  ggtitle("gastrocnemius") +
  ylab("rlog(counts)")

p2 <- DESeq2::rlog(as.matrix(counts.soleus)) %>%
  as.data.frame() %>%
  plotBoxLog(metadata.soleus) + 
  ggtitle("soleus") +
  theme(axis.ticks.y = element_blank(),
        axis.text.y = element_blank()) +
  rremove("ylab")
```


```{r counts.rlog.pdf, echo=FALSE, message=FALSE}
pdf(NULL) # avoiding blank plot created by ggarrange
p <- ggpubr::ggarrange(p1, p2, ncol=2, common.legend = TRUE, legend = "right")
x = dev.off()  # see above
p
```





# most variable genes (top 1000) & Normalization

To gain a first look at the difference between the samples and tissues, the
the 1000 most variable genes (according to the standard deviation) are identified
in both tissues.

For that, a `z-score` is be applied per gene followed by calculating the `sd`
and filtering for the top 1000 genes with the highest values.


z-score:

$\frac{x - mean(x)}{sd(x)}$

```{r zscore}
zscore <- function(M) {
  s <- apply( M,1,sd )      # standard deviation
  µ <- apply( M, 1, mean )  # mean
  M.z <- (M - µ) / s        # zscore
  return(M.z)
}

# ' returning the raw counts of the most variable genes by applying zscore and sd
mostVariableRows <- function(M, ntop=1000) {
  M.z <- zscore(M)
  
  # ordering by sd descending
  sdev <- apply(M.z, 1, sd)
  M.top <- M[order(sdev, decreasing = T)[1:ntop] , ]
  return(M.top)
}

counts.gastroc.top <- mostVariableRows(counts.gastroc)
counts.soleus.top <- mostVariableRows(counts.soleus)
```


## visualization with Heatmaps

To visualize the top 1000 most variable genes (union of both tissues), the zscore
was applied on the counts and then visualized in a heatmap. By default, hierarchical 
clustering is applied.

```{r countsBoth, message=FALSE}
# take union of most variable genes from both tissues
most_variable_genes <- c(rownames(counts.gastroc.top),
                         rownames(counts.soleus.top)) %>%
  unique() # 1383 genes

counts.both.top <- 
  merge(counts.gastroc[most_variable_genes, ],
        counts.soleus[most_variable_genes, ],
        by = 0,
        all = T,
        suffixes = c("_gastrocnemius", "_soleus")) %>%
  tibble::column_to_rownames("Row.names") %>% # 1418 genes (why is this higher?)
  na.omit() # 1332 genes
```


```{r combinedHeatmap}
# using the function from "code/visualizations_functions.R"
chm <- plot_combined_CHM(zscore(counts.both.top), datatype = "zscore", params)
```

```{r heatmapBothSVG, include=FALSE}
svglite::svglite( file.path(plot.dir, "01_heatmap_mvg_both.svg") )
chm
dev.off()
```


## PCA

To the same genes is now a principal component analysis applied.

```{r pca.both}
# using the function from "code/visualizations_functions.R"
p <- plot_combined_pca(counts.both.top, params)
p
```


```{r pca.both_save, include=FALSE}
ggsave(file.path(plot.dir, "01_pca_mvg_both.svg"), p)
ggsave(file.path(plot.dir, "01_pca_mvg_both.png"), p)
```



# cleaning up

To save the counts, a `SummarizedExperiment` object is created.
This has the advantage of separating gene annotations (`rowData`) and sample metadata (`colData`)
from the actual counts (`assay`)

As established in the beginning of this document, the gene annotations of
gastrocnemius and soleus differed. It was assumed this was due to an error in the 
soleus data, since it contained several `NA` values. As the genes (per Ensebml ID)
were the same in both tissues, the gene annotation from gastrocnemius were used
for both tissues. If the difference in the annotations was actually intended, this
section needs to be adjusted.

```{r SummarizedExperiment, include=FALSE}
# ordering data.frames by rowname (ENSMUSG ID)
gastroc <- gastroc[order(rownames(gastroc)),]
soleus <- soleus[order(rownames(soleus)),]

# rowData (all rows which were not filtered out before)
rowdata.gastroc <- gastroc[, 13:21]
rowdata.soleus <- rowdata.gastroc # should be the same I think
rowdata.gastroc$filtered = rownames(rowdata.gastroc) %in% rownames(counts.gastroc)
rowdata.soleus$filtered  = rownames(rowdata.soleus) %in% rownames(counts.soleus)

se.gastroc <- SummarizedExperiment(
  assays = list(
    "gastroc" = as.matrix(gastroc[, 1:12])
  ),
  rowData = rowdata.gastroc,
  colData = metadata.gastroc[, 3, drop = F] # only genotype is available as information
)

se.soleus <- SummarizedExperiment(
  assays = list(
    "soleus" = as.matrix(soleus[, 1:12])
  ),
  rowData = rowdata.soleus,
  colData = metadata.soleus[, 3, drop = F] # only genotype is available as information
)
```


The filtered counts (with zero-counts omitted) can be accessed as following:

```{r SE-filter, eval=FALSE}
counts.preFiltered <- se[rowData(se)$filtered,] %>% assay()
```



```{r saveRObjects, include=FALSE}
# save R Objects
dir.Robjects <- file.path("data", "Robjects")
if (!dir.exists(dir.Robjects)) {
  dir.create(dir.Robjects, recursive = T)
}

saveRDS(se.gastroc, file = file.path(dir.Robjects, "01_se.gastroc.rds"))
saveRDS(se.soleus,  file = file.path(dir.Robjects, "01_se.soleus.rds"))
```
