---
title: "01_FirstAnalysis"
author: "Nick Diercksen"
date: "Last compiled on `r format(Sys.time())`"
output:
  html_notebook:
    code_folding: hide
    toc: yes
    toc_float: yes
    df_print: paged
editor_options: 
  chunk_output_type: inline
params:
  condition_pal: !r c("WT"= "#00C3C6","KO"= "#FF6C67")
  tissue_pal: !r c(gastroc = "orange", soleus="purple")
  heatmap_pal: !r RColorBrewer::brewer.pal(11, "RdYlBu")
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(dplyr)
library(ggplot2)
library(treemap)
library(DESeq2)
library(ggfortify) # autoplot (PCA)
library(ggpubr)

# Heatmaps
# ## Github & ComplexHeatmap ----
if (! "ComplexHeatmap" %in% row.names(installed.packages())) {
  devtools::install_github("jokergoo/ComplexHeatmap", force = TRUE)
  # BiocManager::install("ComplexHeatmap")
}
library(ComplexHeatmap)
library(circlize)
library(RColorBrewer)
library(SummarizedExperiment)
```

# Overview

## Preparing data

```{r}
gastroc <-
  readr::read_tsv("./data/counts/readcount_genename_gastroc.xls",
                  show_col_types = F) %>%
  arrange(gene_id) %>% 
  tibble::column_to_rownames(var = "gene_id")
counts.gastroc <- gastroc[, 1:12]

# TODO: fix annotations: several entries are incomplete whereas the gastroc seems to be complete
soleus <-
  readr::read_tsv("./data/counts/readcount_genename_soleus.xls",
                  show_col_types = F) %>%
  arrange(gene_id) %>% 
  tibble::column_to_rownames(var = "gene_id")
counts.soleus <- soleus[, 1:12]

```

```{r compareAnnotations}
# soleus contains `NA` values, whereas gastroc does not:
anyNA(gastroc) # FALSE
anyNA(soleus)  # TRUE

# rownames (gene_id)
assertthat::are_equal(rownames(gastroc), rownames(soleus)) # TRUE


# COMPARE COMLUMNS:
col_idx <- c(1,10,2,11,3,12,4,13,5,14,6,15,7,16,8,17,9,18)
compare.df <- merge(gastroc[13:21], soleus[13:21], by = 0, suffixes = c(".ga", ".so")) %>% 
  tibble::column_to_rownames("Row.names") %>% 
  .[,col_idx]


# `gene_name`
assertthat::are_equal(gastroc$gene_name, soleus$gene_name) # FALSE

# `gene_start`
assertthat::are_equal(compare.df$gene_start.ga, compare.df$gene_start.so) # FALSE
# filter(compare.df, gene_start.ga != gene_start.so) %>% View() # 504 entries

# `gene_end`
assertthat::are_equal(compare.df$gene_end.ga, compare.df$gene_end.so) # FALSE
# filter(compare.df, gene_end.ga != gene_end.so) %>% View() # 506 entries

# `gene_chr`
assertthat::are_equal(compare.df$gene_chr.ga, compare.df$gene_chr.so) # FALSE
# filter(compare.df, gene_chr.ga != gene_chr.so) %>% View() # 58 entries (in `so` no `chr` available)

# `gene_strand`
assertthat::are_equal(compare.df$gene_strand.ga, compare.df$gene_strand.so) # FALSE
# filter(compare.df, gene_strand.ga != gene_strand.so) %>% View() # 33 entries (all inverse, where 32 are "+" in ga)

# `gene_biotype`
assertthat::are_equal(compare.df$gene_biotype.ga, compare.df$gene_biotype.so) # FALSE
# filter(compare.df, gene_biotype.ga != gene_biotype.so) %>% View() # 155 entries

# `gene_description`
assertthat::are_equal(compare.df$gene_description.ga, compare.df$gene_description.so) # FALSE
# filter(compare.df, gene_description.ga == gene_description.so) %>% View() # 58 entries have no description at all, gastroc has no description at all

# `tf_family`
assertthat::are_equal(compare.df$tf_family.ga, compare.df$tf_family.so) # FALSE
# filter(compare.df, tf_family.ga != tf_family.so) %>% View() # 11096 entries, 

rm(compare.df)
```

Genes for both tissues have the same ensembl id, and have in general the same annotations.
But each annotation (column) has entries which are inconsistent!

Thus maybe only the gene_ids will be used in following analysis. (TODO?)


## Treemap

Both counts have the same genes analysed.
These amount to nearly 54,000 genes, which corresponds to the whole genome of mice. 

```{r gene_biotypes}
gene_biotypes <- table(gastroc$gene_biotype) %>%
  as.data.frame() %>%
  arrange(desc(Freq))

treemap::treemap(gene_biotypes,
                 index = "Var1",
                 vSize = "Freq",
                 title = "Gene biotypes")
```

# Filtering

## Overview

For a broad overview all genes with `rowSums() < 10` will be displayed. This is usually similar to omitting genes with zero counts of more than 30%

```{r overview}
rs.gastroc <- rowSums(counts.gastroc) %>%
  as.data.frame() %>%
  dplyr::rename("rowSums" = ".")
n.omit.gastroc <- sum(rs.gastroc < 10)
title <- paste0("rowSums low expressed genes, gastroc \n",
                n.omit.gastroc, " genes will be omitted (~",
                floor(n.omit.gastroc / nrow(rs.gastroc) * 100), " %)")

p.gastroc <- filter(rs.gastroc, rowSums < 20) %>%
  ggplot(., aes(x=rowSums))+
  geom_histogram(color="black", fill="white")+
  geom_vline(aes(xintercept=10, color="red"),
             linetype="dashed") +
  ggtitle(title) + 
  theme(legend.position = "none")

# ---
rs.soleus <- rowSums(counts.soleus) %>%
  as.data.frame() %>%
  dplyr::rename("rowSums" = ".")
n.omit.soleus <- sum(rs.soleus < 10)
title <- paste0("rowSums low expressed genes, soleus \n",
                n.omit.soleus, " genes will be omitted (~",
                floor(n.omit.gastroc / nrow(rs.gastroc) * 100), " %)")

p.soleus <- filter(rs.soleus, rowSums < 20) %>%
  ggplot(., aes(x=rowSums))+
  geom_histogram(color="black", fill="white")+
  geom_vline(aes(xintercept=10, color="red"),
             linetype="dashed") +
  ggtitle(title)+ 
  theme(legend.position = "none")

p.gastroc
p.soleus

```

## omitting genes (actual filtering)

omitting genes with zero counts in more than 30% of the samples

```{r omit_zero_counts, echo=FALSE}
omitZeroCounts <- function(counts.raw, share=0.3) {
  keep <- (counts.raw != 0) %>%
    rowSums() > round(share * ncol(counts.raw))
  return(counts.raw[keep,])
}

counts.gastroc <- omitZeroCounts(counts.gastroc)
counts.soleus <- omitZeroCounts(counts.soleus)
```



# Quality assessment

## Library sizes

-   too many zero counts
-   different total counts

```{r lib_sizes}
plot.library_size <- function(counts, title="") {
  cs <- colSums(counts) %>%
    as.data.frame() %>%
    dplyr::rename("colsums" = ".") %>%
    tibble::rownames_to_column(var = "sample")
  
  ggplot(cs, aes(x = sample, y = colsums)) +
    geom_bar(stat = "identity") +
    theme(axis.text.x=element_text(color = "black", angle=45, vjust=.8, hjust=0.8)) + 
    ggtitle(paste0("total counts, ", title))
}


plt.libsize.gastroc <- plot.library_size(counts.gastroc, title = "gastroc") +
  ylim(0, 4e07)
plt.libsize.soleus <- plot.library_size(counts.soleus, title = "soleus") +
  ylim(0, 4e07)

ggpubr::ggarrange(plt.libsize.gastroc, plt.libsize.soleus,
                    ncol = 1, nrow = 2)
```

not sure what to make of this, but its there.


### count distribution boxplots

Preparing the sample by applying log2 and loading metadata files

```{r counts.distr.prep, echo=FALSE, message=FALSE}
logcounts.gastroc <- log2(counts.gastroc + 1) # avoiding -Inf with ++
logcounts.soleus <- log2(counts.soleus + 1) # avoiding -Inf with ++

metadata.gastroc <- readr::read_csv2("./data/counts/sample_table_gastroc.csv") %>%
  tibble::column_to_rownames("run")
metadata.soleus <- readr::read_csv2("./data/counts/sample_table_soleus.csv") %>%
  tibble::column_to_rownames("run")
```


Resulting Plots:

```{r counts.distr.plot, message=FALSE}
plotBoxLog <- function(logcounts, metadata) {
  stack(logcounts) %>%
    merge(metadata, by.x = "ind", by.y = "sample_name") %>% 
    ggplot( aes(x=ind, y=values, color=genotype)) +
    scale_color_manual(values = params$condition_pal) +
    geom_boxplot() + 
    ylab("Log2(Counts)") +
    xlab(element_blank()) +
    theme(axis.text.x=element_text(angle=45, vjust=.8, hjust=0.8)) +
    geom_hline(yintercept = median(as.matrix(logcounts)), col="blue" )
}

p1 <- plotBoxLog(logcounts.gastroc, metadata.gastroc) + 
  ggtitle("gastroc")
p2 <- plotBoxLog(logcounts.soleus, metadata.soleus) + 
  ggtitle("soleus") +
  theme(axis.ticks.y = element_blank(),
        axis.text.y = element_blank()) +
  rremove("ylab")
```


```{r counts.distr.pdf, echo=FALSE, message=FALSE}
pdf(NULL) # avoiding blank plot created by ggarrange
p <- ggpubr::ggarrange(p1, p2, ncol=2, common.legend = TRUE, legend = "right")
x = dev.off()  # see above
p
```



### using DESeq2::rlog to normalize

Just for exploration purpose, but the resulting data is not used in the
following analysis.

```{r counts.rlog.plot, message=FALSE}
p1 <- DESeq2::rlog(as.matrix(counts.gastroc)) %>%
  as.data.frame() %>%
  plotBoxLog(metadata.gastroc) + 
  ggtitle("gastroc") +
  ylab("rlog(counts)")

p2 <- DESeq2::rlog(as.matrix(counts.soleus)) %>%
  as.data.frame() %>%
  plotBoxLog(metadata.soleus) + 
  ggtitle("soleus") +
  theme(axis.ticks.y = element_blank(),
        axis.text.y = element_blank()) +
  rremove("ylab")
```


```{r counts.rlog.pdf, echo=FALSE, message=FALSE}
pdf(NULL) # avoiding blank plot created by ggarrange
p <- ggpubr::ggarrange(p1, p2, ncol=2, common.legend = TRUE, legend = "right")
x = dev.off()  # see above
p
```


# most variable genes (top 1000) & Normalization
To obtain the most variable genes, a `z-score` will be applied per gene to then
take the `sd` and filter for the top 1000


z-score:

$\frac{x - mean(x)}{sd(x)}$

```{r zscore}
zscore <- function(M) {
  s <- apply( M,1,sd )      # standard deviation
  µ <- apply( M, 1, mean )  # mean
  M.z <- (M - µ) / s        # zscore
  return(M.z)
}

# ' returning the raw counts of the most variable genes by applying zscore and sd
mostVariableRows <- function(M, ntop=1000) {
  M.z <- zscore(M)
  
  # ordering by sd descending
  sdev <- apply(M.z, 1, sd)
  M.top <- M[order(sdev, decreasing = T)[1:ntop] , ]
  return(M.top)
}

counts.gastroc.top <- mostVariableRows(counts.gastroc)
counts.soleus.top <- mostVariableRows(counts.soleus)
```


## visualization with Heatmaps
To visualize the top 1000 most variable genes, I tried to generate some Heatmaps. (This time not using DESeq2::rlog, but just `log2(counts)`. Using the raw counts
I could try to just apply `log2` to all counts and then create a heatmap.

### gastroc

```{r fig.width=4, fig.height=4, message=FALSE}

# ' returning a CHM with the z-scored counts
plotCHM <- function(counts) {
  # http://colorbrewer2.org/#type=sequential&scheme=BuGn&n=3
  morecols <- colorRampPalette(params$heatmap_pal)
  
  chm <- Heatmap(
    zscore(counts),
    row_title = "genes",
    name = "zscore",
    show_row_names = FALSE,
    column_title = "samples",
    col = rev(morecols(50)),
    column_title_side = "bottom"
  )
  return(chm)
}

plotCHM(counts.gastroc.top)
```


### soleus

```{r fig.width=4, fig.height=4, message=FALSE}
plotCHM(counts.soleus.top)
```



# PCA

## gastroc

the pca is performed on the z-scored data
=> Thus not using the `scale.=T` argument?

```{r pca.gastroc, message=FALSE}
pca <- prcomp(t(counts.gastroc.top), .scale=T)

pca.data <- data.frame(Sample = rownames(pca$x),
                       X = pca$x[, 1],
                       Y = pca$x[, 2]) %>%
  mutate(type = substr(Sample, 1,2))

plt <-
  autoplot(
    pca,
    data = pca.data,
    colour = 'type',
    label.show.legend = FALSE
  ) +
  ggtitle("PCA gastroc")+
  theme_bw()
ggsave(filename = "./plots/04_pca_gastroc_filtered.svg", plt)
plt
```

## soleus

```{r pca.soleus, message=FALSE}
pca <- prcomp(t(counts.soleus.top), scale. = T)

pca.data <- data.frame(Sample = rownames(pca$x),
                       X = pca$x[, 1],
                       Y = pca$x[, 2]) %>%
  mutate(type = substr(Sample, 1,2))

plt <-
  autoplot(
    pca,
    data = pca.data,
    colour = 'type',
    shape = 17,
    label.show.legend = FALSE
  ) +
  ggtitle("PCA soleus")+
  theme_bw()
ggsave(filename = "./plots/04_pca_soleus_filtered.svg", plt)
plt
```

## both tissues
Here also taking the top 1000 most variable genes of each tissue and merging them.

```{r pca.both, message=FALSE}
# combine both count tables
counts.both <-
  merge(counts.gastroc.top,
        counts.soleus.top,
        by = 0,
        suffixes = c("_gastroc", "_soleus")) %>%
  tibble::column_to_rownames("Row.names")

pca <- prcomp(t(counts.both), scale. = T)

pca.data <- data.frame(Sample = rownames(pca$x),
                       X = pca$x[, 1],
                       Y = pca$x[, 2]) %>%
  mutate(type = substr(Sample, 1, 2),
         tissue = stringr::str_extract(Sample,"[:alpha:]+$"))

plt <-
  autoplot(
    pca,
    data = pca.data,
    colour = 'type',
    shape = 'tissue',
    label.show.legend = FALSE
  ) +
  ggtitle("PCA both tissues")+
  theme_bw()
ggsave(filename = "./plots/04_pca_both_tissues_filtered.svg", plt)
plt
```


# cleaning up

## SE

! important: here the rowdata of the gastroc is used as the rowdata of the soleus.
TODO: this should be changed I believe, maybe just use here the ensembl ids and annotate them here new, so that the annotations are the same for both tissues?
Or is it actually intended, that the annotations differ?

```{r SummarizedExperiment}
# ordering data.frames by rowname (ENSMUSG ID)
gastroc <- gastroc[order(rownames(gastroc)),]
soleus <- soleus[order(rownames(soleus)),]

# rowData (all rows which were not filtered out before)
rowdata.gastroc <- gastroc[, 13:21]
rowdata.soleus <- rowdata.gastroc # should be the same I think
rowdata.gastroc$filtered = rownames(rowdata.gastroc) %in% rownames(counts.gastroc)
rowdata.soleus$filtered  = rownames(rowdata.soleus) %in% rownames(counts.soleus)

se.gastroc <- SummarizedExperiment(
  assays = list(
    "gastroc" = as.matrix(gastroc[, 1:12])
  ),
  rowData = rowdata.gastroc,
  colData = metadata.gastroc[, 3, drop = F] # only genotype is available as information
)

se.soleus <- SummarizedExperiment(
  assays = list(
    "soleus" = as.matrix(soleus[, 1:12])
  ),
  rowData = rowdata.soleus,
  colData = metadata.soleus[, 3, drop = F] # only genotype is available as information
)
```

the filtered assays can be accessed as following:
```{r SE-filter, eval=FALSE}
counts.preFiltered <- se[rowData(se)$filtered,] %>% assay()
```

## save R Objects

```{r}
saveRDS(se.gastroc,             file = "./data/Robjects/01_se.gastroc.rds")
saveRDS(se.soleus,             file = "./data/Robjects/01_se.soleus.rds")

# (redundant)
saveRDS(counts.gastroc, file = "./data/Robjects/counts.gastroc.rds")
saveRDS(counts.soleus,  file = "./data/Robjects/counts.soleus.rds")
```