---
title: "02 Annotation"
author: "Nick Diercksen"
output:
  html_notebook:
    toc: yes
    toc_float: yes
    df_print: paged
editor_options: 
  chunk_output_type: inline
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(dplyr)
library(biomaRt)  # interface to a collection of databases
library(stringr)
library(SummarizedExperiment)
```

# About
This step is a preparation for the following pathway enrichment (gene set 
enrichment with `fgsea`). To link the genes with the pathways the "entrezgene_id"
is needed. Thus in this step, the `biomaRt` interface is used to access the 
corresponding DBs.

# read in data

```{r}
se <- readRDS("./data/Robjects/01_se.rds")
```


# setting up connection to DB

## Select BioMart database and dataset

```{r connect}
# view the available databases
listMarts()
## set up connection to ensembl database
ensembl=useMart("ENSEMBL_MART_ENSEMBL")
# list the available datasets (species)
listDatasets(ensembl) %>% 
    filter(str_detect(description, "Mouse"))
# specify a data set to use
ensembl = useDataset("mmusculus_gene_ensembl", mart=ensembl)
```

## Query the database

### explore filter types

```{r queryBioMart, message=F}
# check the available "filters" - things you can filter for
listFilters(ensembl) %>% 
    filter(str_detect(name, "ensembl"))
# check the available "attributes" - things you can retrieve
listAttributes(ensembl) %>% head(40)
```




*  multiple Entrez IDs for a single Ensembl 


## Retrieve full annotation

* **values** - the Ensembl Gene IDs from our DE results table
* **attributes** - columns 

```{r}
# 
filterValues <- rownames(se)
filterType <- "ensembl_gene_id"
attributeNames <- c('ensembl_gene_id',
                    'entrezgene_id',
                    'external_gene_name',
                    'description',
                    'gene_biotype',
                    'chromosome_name',
                    'start_position',
                    'end_position',
                    'strand')
# run the query
annot <- getBM(attributes=attributeNames,
               filters = filterType,
               values = filterValues,
               mart = ensembl)
```


duplicate ids:
```{r}
sum(duplicated(annot$ensembl_gene_id))
```


missing genes:
```{r}
missingGenes <- !rownames(se) %in% annot$ensembl_gene_id
rownames(se)[missingGenes] %>% head(20)
```


# consistency in the metadata (sanity checks)

the metadata from the genes originates from tow sources:

* already provided with the available counts
* from biomart (Ensembl)

The question is if both datasets are consistent with the data they provide.
Inconsistencies can point to possible mistakes done when annotating.

So before merging the new annotation data into the `se` to get the "entrezgene_id"
to later do the pathway enrichment.

```{r}
df <- merge(annot, rowData(se), by.x = 1, by.y = 0) %>%
  as.data.frame()
```

## biotypes:
```{r}
cols <- c("ensembl_gene_id","gene_biotype.x", "gene_biotype.y")
dif_biotypes <- !df$gene_biotype.x == df$gene_biotype.y
print(paste0("There are ", sum(dif_biotypes), " genes annotated with different biotypes"))
df[(dif_biotypes),cols] %>% head(20)
```

## gene names
```{r}
cols <- c("ensembl_gene_id", "external_gene_name", "gene_name")
dif_names <- !df$external_gene_name == df$gene_name
print(paste0("There are ", sum(dif_names), " genes annotated with different gene names"))
df[(dif_names), cols] %>% head(20)
```

## chromosome
```{r}
cols <- c("ensembl_gene_id", "chromosome_name", "gene_chr", "external_gene_name", "gene_name")
df[(!df$chromosome_name == df$gene_chr), cols] %>% head(20)
```

Here we "only" have 2 genes annotated with different chromosomes, but this might
be one of the more severe differences!

**ENSMUSG00000065387** \\



## gene positions
-> start
```{r}
dif_start <- df$start_position - df$gene_start
head(dif_start, 20)
```

total genes with the same start position
```{r}
sum(dif_start == 0)
```

```{r}
hist(abs(dif_start), breaks = 100)
```


-> end
```{r}
dif_end <- df$end_position - df$gene_end
head(dif_end, 20)
```


total genes with the same end position

```{r}
sum(dif_end == 0)
```

```{r}
hist(abs(dif_end), breaks = 100)
```

## Discussion
On the first glance, there are several inconsistencies with the two annotation
sources.

Biotypes: \\
About a fifth of the genes is annotated with a different biotype. This has to be
looked at more specifically though, since this could still be viable in the 
biological sense.

gene names: \\
Not as many differences as the biotypes, but most of them seem to be synonyms:

* http://www.ensembl.org/Mus_musculus/Gene/Summary?db=core;g=ENSMUSG00000002280;r=17:25992750-26002306
* http://www.ensembl.org/Mus_musculus/Gene/Summary?db=core;g=ENSMUSG00000003527;r=16:17718573-17729212

chromosomes: \\
This

Gene length was not compared, since no field was available and subtracting the
start from the end position would not make sense, since introns would probably 
get counted as well.

# save Data

```{r}
save(annot, file = "./data/Robjects/02_annot.RData")
```



# Questions
