---
title: "02_DEA"
author: "Nick Diercksen"
output:
  html_notebook:
    code_folding: hide
    toc: yes
    toc_float: yes
    df_print: paged
  html_document:
    toc: yes
    df_print: paged
editor_options:
  chunk_output_type: inline
params:
  FCutoff:
    label: cutoff for the log2FoldChange
    value: 1
  pCutoff:
    label: "cutoff for the adjusted p-Value"
    value: 0.01
  condition_pal: !r c("WT"= "#00C3C6","KO"= "#FF6C67")
  tissue_pal: !r c("gastroc" = "orange", "soleus"="purple")
  regulated_pal: !r list("upregulated" = 'royalblue', "downregulated" = 'red', "insignificant" = 'gray')
  heatmap_pal: !r rev(RColorBrewer::brewer.pal(11, "RdYlBu"))
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(dplyr)
library(ggplot2)
library(DESeq2)
library(SummarizedExperiment)


if (!require("EnhancedVolcano", quietly = TRUE))
  BiocManager::install("EnhancedVolcano")
if (!require("eulerr", quietly = TRUE))
  install.packages("eulerr")
library(EnhancedVolcano)
library(eulerr)
library(RColorBrewer)


if (!require("ashr", quietly = TRUE))
  install.packages("ashr") # needed for lfcShring

```


```{r plottingSettings, include=FALSE}
# for custom CHM and PCA plotting functions
source(file.path("code", "visualization_functions.R"))

ggplot2::theme_set(theme_bw())

plot.dir <- file.path("plots", "02 DEA")
```


In this Analysis step, the Differential Expression Analysis (DEA) is performed
using the `DESeq2` package.

After performing the DEA, the quality of the data is investigated with a look at
the dispersion estimates, size factors and MA plots.

Goal is to find the most significant differential expressed genes. These will be
visualized using Volcano plots. To compare the two tissues, the shared significant
genes are plotted against each other and a heatmap with hierarchical clustering
is performed on all significant genes of both tissues combined followed by a
principle component analysis.


# loading data

For the DESeq2 analysis, the raw, prefiltered counts will be used.

```{r loadRDS}
se.gastroc <- readRDS("./data/Robjects/01_se.gastroc.rds")
se.soleus <- readRDS("./data/Robjects/01_se.soleus.rds")

counts.gastroc <- se.gastroc[rowData(se.gastroc)$filtered,] %>% assay()
counts.soleus <- se.soleus[rowData(se.soleus)$filtered,] %>% assay()
metadata <- colData(se.gastroc)
```

# preparing DDS objects

The DDS objects are created for both tissues separately to investigate the differential expression in regard to the genotype (condition of KO or WT).

```{r prepDDS, message=FALSE, warning=FALSE}
# changing to factor (needs DESeq)
metadata$genotype <- as.factor(metadata$genotype)

createDDSObject <- function(counts, metadata) {
  # select sample columns
  reorder_index <- match(rownames(metadata), colnames(counts))
  counts <- counts[,reorder_index] %>% 
    apply(c(1,2), as.integer) # DESeq need normal counts (integer) as input
  
  # Check metadata consistency
  all(rownames(metadata) %in% colnames(counts)) %>%
    assertthat::assert_that(., msg = "metadata and count table do not match")
  
  ## DESeq2 object
  dds <- DESeqDataSetFromMatrix(countData = counts,
                                colData = metadata,
                                design = ~ genotype)
  
  return( DESeq(dds) )
}

# creating DESeq Objects
dds.gastroc <- createDDSObject(counts.gastroc, metadata)
dds.soleus <- createDDSObject(counts.soleus, metadata)
```



# DEA

Performing the Differential Expression Analysis and gathering the results from the DDS objects.

```{r results}
res.gastroc <- results(dds.gastroc, alpha = params$pCutoff, contrast = c("genotype", "KO", "WT"))
res.soleus <- results(dds.soleus, alpha = params$pCutoff, contrast = c("genotype", "KO", "WT"))
```


Also applying lfcShrink for comparison in volcano plots. Besides that it's application
was not further used.

```{r lfc}
# lfcShrink is not applied by default

# library(ashr) # using instead of "apeglm", because https://www.bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#extended-section-on-shrinkage-estimators

resLFC.gastroc <- lfcShrink(dds.gastroc, contrast = c("genotype", "KO", "WT"), type = "ashr")
resLFC.soleus <- lfcShrink(dds.soleus, contrast = c("genotype", "KO", "WT"), type = "ashr")
```


# Exploration and Quality Control


## count outlier detection with cooks distance

Filtering outliers using the cooks distance is done automatically and p-values are set to `NA`.
Changing (lowering) the `minReplicatesForReplace` will replace these flagged values.

DESeq2 flags these outliers and removes the corresponding genes completely, if below 7 samples (above would replace the values)


```{r}
# assuming all NA p-values are the ones filtered out using the cooks distance

print_outliers <- function(res, se){
  # all genes with a `NA` value are exactly those 5 which were also detected by the cooks distance:
  outlier_ids <- res %>% 
    as.data.frame() %>% 
    filter(is.na(pvalue)) %>% 
    rownames()
  
  # a different approach would be finding the genes which have a cooks distance
  # above the cutoff:
  # cooksCutoff <- qf(.99, 2,10)
  # outlier_ids <- which(mcols(dds)$maxCooks > cooksCutoff) %>% names()
  
  # getting counts and metadata
  outlier_metadata.df <- rowData(se) %>%
    as.data.frame() %>%
    select(gene_name, gene_biotype)
  
  counts.df <- assay(se)
  
  # merging df
  outliers.df <-
    merge(outlier_metadata.df, counts.df[outlier_ids,], by = 0)
  return(outliers.df)
}

outliers_gastroc.df <- print_outliers(res.gastroc, se.gastroc) %>% tibble::column_to_rownames(var = "Row.names")
write.csv(outliers_gastroc.df, file = "./code/analysis/02_outliers_gastroc.csv")
outliers_gastroc.df %>% knitr::kable()
print_outliers(res.soleus, se.soleus) %>% knitr::kable()
```


## size factors
```{r size_factors}
sizeFactors(dds.gastroc)
sizeFactors(dds.soleus)
```



## Dispersion Estimates

### gastroc

```{r dispEst.gastroc}
plotDispEsts(dds.gastroc)
```

### soleus

```{r dispEst.soleus}
plotDispEsts(dds.soleus)
```


## MA-Plots

### gastroc

```{r ma.gastroc, message=FALSE, warning=FALSE}
plotMA(res.gastroc, ylim = c(-6,6))
```

```{r maLFC.gastroc, message=FALSE, warning=FALSE, include=FALSE, eval=FALSE}
# log fold change shrinked:
plotMA(resLFC.gastroc)
```

### soleus

```{r ma.soleus, message=FALSE, warning=FALSE}
plotMA(res.soleus, ylim = c(-6,6))
```

```{r maLFC.soleus, message=FALSE, warning=FALSE, include=FALSE, eval=FALSE}
# log fold change shrinked:
plotMA(resLFC.soleus)
```


# top significant diff. genes

```{r topDiffGenes.df}
topGenes.gastroc <- as.data.frame(res.gastroc) %>%
  tibble::rownames_to_column("GeneID") %>%
  top_n(100, wt = -padj) %>%
  arrange(padj)

topGenes.soleus <- as.data.frame(res.soleus) %>%
  tibble::rownames_to_column("GeneID") %>%
  top_n(100, wt = -padj) %>%
  arrange(padj)

knitr::kable(head(topGenes.gastroc), caption = "gastroc")
knitr::kable(head(topGenes.soleus), caption = "soleus")
```

## p-values

```{r pValuesHist}
# TODO: use ggplot for unified look of plots
hist(res.gastroc$pvalue, main = "gastroc")
hist(res.soleus$pvalue, main = "soleus")
```


## Volcano Plot
```{r volcanoPlot, warning=FALSE, fig.width = 9, fig.height = 9, fig.retina = 2, dpi = 100}
volcanoPlot <- function(result, se, pCutoff = 0.01, FCutoff = 1, tissue = character()) {
  gene_names <-
    rowData(se)[rownames(result), c("gene_name"), drop = F]
  results.df <- result %>%
    as.data.frame() %>%
    dplyr::arrange(padj)
  
  # top 10 gene labels for respectively up and down regulation
  labs.up <- results.df[results.df$log2FoldChange > FCutoff, ] %>%
    rownames() %>% .[1:10] %>% gene_names[., c("gene_name")]
  labs.down <- results.df[results.df$log2FoldChange < -FCutoff, ] %>%
    rownames() %>% .[1:10] %>% gene_names[., c("gene_name")]
  selectLab <- c(labs.up, labs.down, "Nfe2l1") %>% unique() # always including "Nfe2l1"
  
  # custom colors:
  keyvals <- ifelse(
    result$log2FoldChange > FCutoff &
      result$padj < pCutoff,
    params$regulated_pal$upregulated,
    ifelse(
      result$log2FoldChange < -FCutoff &
        result$padj < pCutoff,
      params$regulated_pal$downregulated,
      params$regulated_pal$insignificant
    )
  )

  keyvals[is.na(keyvals)] <- params$regulated_pal$insignificant
  names(keyvals)[keyvals == params$regulated_pal$upregulated] <- 'up regulated'
  names(keyvals)[keyvals == params$regulated_pal$insignificant] <- 'nonsignificant'
  names(keyvals)[keyvals == params$regulated_pal$downregulated] <- 'down regulated'
  
  vlc.plt <- EnhancedVolcano(
    result,
    x = 'log2FoldChange',
    y = 'padj',
    title = 'KO vs WT: Nfe2l1 knockout',
    subtitle = ifelse(isEmpty(tissue), "", paste0('tissue: ', tissue)),
    caption = "",
    ylab = expression(paste(-Log[10], p[adj])),
    pCutoff = pCutoff,
    FCcutoff = FCutoff,
    legendPosition = 'right',
    pointSize = 2,
    colCustom = keyvals,
    lab = gene_names$gene_name,
    selectLab = selectLab,
    labSize = 3,
    boxedLabels = TRUE,
    drawConnectors = TRUE,
    max.overlaps = Inf
  )
  
  return(vlc.plt)
}
```


```{r volcano, warning=FALSE, fig.width = 9, fig.height = 9, fig.retina = 2, dpi = 100}
p <- volcanoPlot(res.gastroc, se.gastroc, pCutoff = params$pCutoff, FCutoff = params$FCutoff, tissue = "gastrocnemius")
ggsave(file.path(plot.dir, "02_volcano_gastroc.svg"), p)
ggsave(file.path(plot.dir, "02_volcano_gastroc.png"), p)
p
p <- volcanoPlot(res.soleus, se.soleus, pCutoff = params$pCutoff, FCutoff = params$FCutoff, tissue = "soleus")
ggsave(file.path(plot.dir, "02_volcano_soleus.svg"), p)
ggsave(file.path(plot.dir, "02_volcano_soleus.png"), p)
p
```



```{r volcanoLFC, fig.width = 9, fig.height = 9, fig.retina = 2, dpi = 100}

p <- volcanoPlot(resLFC.gastroc, se.gastroc, pCutoff = params$pCutoff, FCutoff = params$FCutoff, tissue = "gastrocnemius")
ggsave(file.path(plot.dir, "02_volcano_gastroc_LFC.svg"), p)
ggsave(file.path(plot.dir, "02_volcano_gastroc_LFC.png"), p)
p
p <- volcanoPlot(resLFC.soleus, se.soleus, pCutoff = params$pCutoff, FCutoff = params$FCutoff, tissue = "soleus")
ggsave(file.path(plot.dir, "02_volcano_soleus_LFC.svg"), p)
ggsave(file.path(plot.dir, "02_volcano_soleus_LFC.png"), p)
p
```



## scatter-plot: most differential genes, both tissues

using the Wald-test `stat` from the DESeq2 result and filtering on the set
`FCutoff=``r params$FCutoff` and `pCutoff=``r params$pCutoff`


```{r DE_genes_both}
apply_cutoffs <- function(deseq.result, colname="stat", FCutoff, pCutoff) {
  res.filtered <- deseq.result %>%
    data.frame() %>%
    filter(padj < pCutoff,
           log2FoldChange > FCutoff | log2FoldChange < -FCutoff) %>%
    dplyr::rename(!!colname := stat) %>%
    dplyr::select(!!colname)
  return(res.filtered)
}

gastroc_res.filtered <- apply_cutoffs(res.gastroc, colname="gastroc", params$FCutoff, params$pCutoff)
soleus_res.filtered  <- apply_cutoffs(res.soleus,  colname="soleus",  params$FCutoff, params$pCutoff)

gene_names <- rowData(se.gastroc) %>% as.data.frame() %>% 
  dplyr::select(gene_name)

# combining Wald-Test data from both tissues and ordering in quadrants
res.combined <- merge(gastroc_res.filtered,
                      soleus_res.filtered,
                      by = 0) %>%
  mutate(diff.exp = case_when(
    gastroc < 0 & soleus < 0 ~ "both down",
    gastroc > 0 & soleus > 0 ~ "both up",
    gastroc < 0 & soleus > 0 ~ "gastrocnemius down,\n soleus up",
    gastroc > 0 & soleus < 0 ~ "gastrocnemus up,\n soleus down",
    TRUE ~ "different"
  )) %>% 
  merge(gene_names, by.x="Row.names", by.y=0)
```


plotting the filtered (DE) genes.

```{r scatterDEA, warning=FALSE, fig.width = 9, fig.height = 9, fig.retina = 2, dpi = 100}
# removing all gene_names except the top_n_genes (sum of absolute Wald-test numbers)
top_n_genes <- 10
top_labels <- res.combined %>%
  group_by(diff.exp) %>% 
  arrange(desc(abs(gastroc) + abs(soleus))) %>%
  filter(row_number() %in% c(1:top_n_genes)) %>% 
  ungroup() %>% 
  .$gene_name

res.combined <- res.combined %>% 
  mutate(gene_name = ifelse(gene_name %in% top_labels, gene_name, ""))

# final plot
p <- ggplot(res.combined, aes(x = gastroc, y = soleus, label = gene_name)) +
  geom_vline(xintercept = 0) + 
  geom_hline(yintercept = 0) + 
  geom_point(aes(color = diff.exp)) +
  # scale_color_manual(values = c("red", "chartreuse1", "bisque", "royalblue")) +
  labs(x = "gastrocnemius", y = "soleus", color = "significantly\ndifferentially\nexpressed") +
  ggrepel::geom_label_repel(max.overlaps = 15) + 
  ggtitle(label = "DEA: t-statistics (Wald test)") +
  theme_bw() +
  theme(
    axis.title = element_text(size = 15),
    axis.text = element_text(size = 15),
    title = element_text(size = 15),
    legend.text = element_text(size = 14),
    legend.title = element_blank(),
    legend.position = "bottom"
  ) + 
  coord_fixed()

p
```

```{r scatterDEA_save, include=FALSE}
ggsave(file.path(plot.dir, "02_dea_scatter.svg"), p)
ggsave(file.path(plot.dir, "02_dea_scatter.png"), p)
```


### barplot

```{r dea_scatter_barplot}
p <- ggplot(res.combined, aes(x = diff.exp)) +
  geom_bar(aes(fill = diff.exp)) +
  theme_bw() +
  theme(legend.position = "none",
    axis.title = element_text(size = 15),
    axis.text = element_text(size = 12),
    axis.title.x = element_blank()
        )
p
```


```{r dea_scatter_barplot_save, include=FALSE}
ggsave(file.path(plot.dir, "02_dea_scatter_barplot.svg"), p)
ggsave(file.path(plot.dir, "02_dea_scatter_barplot.png"), p)
```


```{r include=FALSE}
# obtain gene counts for the respective groups for visualizing
sign_gene_stats <- list(
  "all genes" = nrow(gene_names),
  "gastrocnemius" = nrow(gastroc_res.filtered),
  "soleus" = nrow(soleus_res.filtered),
  "shared_sig_genes" = nrow(res.combined),
  "both up" = sum(res.combined$diff.exp == "both up"),
  "both down" = sum(res.combined$diff.exp == "both down"),
  "ga up, sol down" = sum(res.combined$diff.exp == "ga up, sol down"),
  "ga down, sol up" = sum(res.combined$diff.exp == "ga down, sol up")
)
```


## boxplots: significant genes in both tissues

```{r boxplotSignGenes, message=FALSE, warning=FALSE}
boxplot.genes <- function(counts.df, title ="", FCutoff = 1) {
  
  ggplot(counts.df,
         aes(
           # x = as.factor(gene_name),
           x = reorder(gene_name,ensembl),
           y = normalized_counts,
           fill = genotype
         )) +
    geom_dotplot(
      binaxis = 'y',
      stackdir = 'center',
      dotsize = 0.3,
      position = position_dodge(0.8),
      fill = "black"
    ) +
    geom_boxplot(outlier.size = 0.3) +
    scale_y_log10() +
    xlab("Genes") +
    ylab("Normalized Counts") +
    scale_fill_manual(values = params$condition_pal) +
    ggtitle(title) +
    theme_bw() +
    theme(plot.title = element_text(hjust = 0.5)) +
    theme(axis.text.x = element_text(
      angle = 45,
      hjust = 1,
      face = setBold(counts.df$gene_name, c("Nfe2l1"))
    ))
}

prepare_counts <- function(gene_ids, dds, se, reorder_genes = F) {
  counts.top <- counts(dds, normalized = T)[gene_ids, ]
  metadata <- colData(se) %>% as.data.frame()
  gene_names <-
    rowData(se)[, c("gene_name"), drop = F] %>% as.data.frame()
  
  # the prepared counts df for the plot
  counts.plt <-
    data.frame(counts.top) %>%
    tibble::rownames_to_column(var = "ensembl") %>%
    tidyr::gather(key = "samplename",
                  value = "normalized_counts", 2:13) %>%
    merge(metadata, by.x = "samplename", by.y = 0) %>%
    merge(gene_names, by.x = "ensembl", by.y = 0)  %>%
    mutate(genotype = factor(genotype)) %>%
    mutate(genotype = relevel(genotype, "WT")) # "WT" needs to be displayed before "KO"
    
    if (reorder_genes) {
      counts.plt <- counts.plt %>%
        mutate(
          gene_name = forcats::fct_reorder(gene_name, normalized_counts, .desc = T),
          ensembl = forcats::fct_reorder(ensembl, normalized_counts, .desc = T)
        )
    } else {
      idx = match(counts.plt$ensembl, gene_ids)
      counts.plt <- counts.plt[order(idx),] %>%
        mutate(gene_name = factor(gene_name))
      
      counts.plt$ensembl <-
        factor(counts.plt$ensembl, levels = gene_ids, ordered = TRUE)
      counts.plt$gene_name <- factor(counts.plt$gene_name)

    }
  return(counts.plt)
}


# helper function to set the font of a label on the axis to bold
setBold <- function(src, special_labs) {
  # source: https://stackoverflow.com/questions/39694490/highlighting-individual-axis-labels-in-bold-using-ggplot2
  if (!is.factor(src))
    src <- factor(src)
  src_levels <- base::levels(src)
  brave <- special_labs %in% src_levels
  b_vec <- rep("plain", length(src_levels))
  if (all(brave)) {
    b_pos <- purrr::map_int(special_labs, ~ which(. == src_levels))
    b_vec[b_pos] <- "bold"
    b_vec
  } else {
    message("setBold: no matching element found")
  }
  return(b_vec)
}

# actual plot call
for (group in unique(res.combined$diff.exp)) {
  gene_ids <- filter(res.combined, diff.exp == group)$`Row.names`
  prep_counts.gastroc <- prepare_counts(gene_ids, dds.gastroc, se.gastroc, reorder_genes = T)
  prep_counts.soleus <-
    prepare_counts(levels(prep_counts.gastroc$ensembl), dds.soleus, se.soleus, reorder_genes = T)
  gene_ids <- levels(prep_counts.gastroc$ensembl)
  
  p.g <- boxplot.genes(
    counts.df = prep_counts.gastroc,
    title = paste0("Sign. regulated genes: ", group, "\n tissue: ", "gastrocnemius")
  ) #%>% print()
  
  p.s <- boxplot.genes(
    counts.df = prep_counts.soleus,
    title = paste0("\n tissue: ", "soleus")
  ) #%>% print()
  
  ggpubr::ggarrange(p.g, p.s, ncol = 1, nrow = 2)
}
```


## boxplots: top 20 significant genes

```{r topDiffGenes}
setBold <- function(src, special_labs) {
  # source: https://stackoverflow.com/questions/39694490/highlighting-individual-axis-labels-in-bold-using-ggplot2
  if (!is.factor(src))
    src <- factor(src)
  src_levels <- base::levels(src)
  brave <- special_labs %in% src_levels
  b_vec <- rep("plain", length(src_levels))
  if (all(brave)) {
    b_pos <- purrr::map_int(special_labs, ~ which(. == src_levels))
    b_vec[b_pos] <- "bold"
    b_vec
  } else {
    message("setBold: no matching element found")
  }
  return(b_vec)
}

getTopExpressedEnsemblNames <- function(result,
                                        FCutoff,
                                        n,
                                        up = T) {
  .filter <- ifelse(up, `>`, `<`)
  
  subset(result, .filter(log2FoldChange, FCutoff)) %>%
    data.frame() %>%
    filter(baseMean > 100) %>%
    arrange(padj) %>%
    .[1:n, ] %>%
    rownames()
}

boxplot.top <- function(result, dds, se, upregulated=TRUE, n = 20, FCutoff = 1, tissue ="") {
  # getting the top n regulated genes
  names.top <- getTopExpressedEnsemblNames(result, FCutoff=FCutoff, n=n, up = upregulated)
  
  counts.top <- counts(dds, normalized=T)[names.top, ]
  metadata <- colData(se) %>% as.data.frame()
  gene_names <- rowData(se)[, c("gene_name"), drop = F] %>% as.data.frame()
  
  counts.plt <-
    data.frame(counts.top) %>%
    tibble::rownames_to_column(var = "ensembl") %>%
    tidyr::gather(key = "samplename",
           value = "normalized_counts", 2:13) %>%
    merge(metadata, by.x="samplename", by.y=0) %>%
    merge(gene_names, by.x="ensembl", by.y=0)  %>%
    mutate(genotype = factor(genotype)) %>%
    mutate(genotype = relevel(genotype, "WT"))  %>% # "WT" needs to be displayed before "KO"
    mutate(gene_name = forcats::fct_reorder(gene_name, normalized_counts, .desc = T))
  
  direction <- ifelse(upregulated, "up", "down")
  
  plot <- ggplot(counts.plt,
         aes(
           x = as.factor(gene_name),
           y = normalized_counts,
           fill = genotype
         )) +
    # geom_dotplot(
    #   binaxis = 'y',
    #   stackdir = 'center',
    #   dotsize = 0.2,
    #   binpositions = "all",
    #   stackgroups = T,
    #   position = position_jitter(-0.3),
    #   fill = "black"
    # ) +
    geom_boxplot(
      outlier.size = 0.2,
      position = position_dodge(0.5),
      outlier.shape = NA,
      ) +
    scale_y_log10(limits = c(1, 1000000)) +
    xlab("Genes") +
    ylab("Normalized Counts") +
    scale_fill_manual(values = params$condition_pal) +
    # ggtitle(paste0("Top ", n, " ", direction, "-regulated Genes\n tissue: ", tissue)) +
    theme_bw() +
    # theme(plot.title = element_text(hjust = 0.5)) +
    theme(axis.text.x = element_text(
      angle = 45,
      hjust = 1,
      face = setBold(counts.plt$gene_name, c("Nfe2l1"))
    ))
  return(plot)
}
# boxplot.top(res.gastroc, dds.gastroc, se.gastroc, upregulated = T, tissue = "gastrocnemius")
```

### gastroc

```{r topDiffBox.gastroc, warning=FALSE, message=FALSE}
bp.gastroc.up <- boxplot.top(res.gastroc, dds.gastroc, se.gastroc, upregulated = T, tissue = "gastrocnemius")
bp.gastroc.up + theme(legend.position = "none", axis.title.x = element_blank())
bp.gastroc.down <- boxplot.top(res.gastroc, dds.gastroc, se.gastroc, upregulated = F, tissue = "gastrocnemius")
bp.gastroc.down + theme(legend.position = "none")
```

### soleus

```{r topDiffBox.soleus, warning=FALSE, message=FALSE}
bp.soleus.up <- boxplot.top(res.soleus, dds.soleus, se.soleus, upregulated = T, tissue = "soleus")
bp.soleus.up + theme(axis.title = element_blank())
bp.soleus.down <- boxplot.top(res.soleus, dds.soleus, se.soleus, upregulated = F, tissue = "soleus")
bp.soleus.down + theme(legend.position = "none", axis.title.y = element_blank())
```


```{r include=FALSE, eval=FALSE}
# Set the plot dimensions and scales
plot_scale <- 0.9  # Adjust this value to resize the plots
plot_width <- 6  # Adjust this value to set the width of each plot
plot_height <- 4  # Adjust this value to set the height of each plot


ggarrange(
  bp.gastroc.up,
  bp.soleus.up + theme(axis.text.y = element_blank()),
  bp.gastroc.down,
  bp.soleus.down + theme(axis.text.y = element_blank()),
  ncol = 2, nrow= 2,
  common.legend = TRUE,
  legend = "bottom",
  widths = c(plot_width, plot_width),
  heights = c(plot_height, plot_height),
  hjust = 0.5,
  vjust = 0.5,
  scale = plot_scale
)

combined_plot <- ggarrange(
  bp.gastroc.up, bp.soleus.up + theme(axis.text.y = element_blank()), bp.gastroc.down, bp.soleus.down + theme(axis.text.y = element_blank()),
  nrow = 2, ncol = 2,
  common.legend = TRUE,
  legend = "bottom",
  widths = c(plot_width, plot_width),
  heights = c(plot_height, plot_height),
  hjust = 0.5, vjust = 0.5,
  scale = plot_scale
) 
combined_plot


annotate_figure(combined_plot,
                top = text_grob("Visualizing len", color = "red", face = "bold", size = 14),
                bottom = text_grob("Data source: \n ToothGrowth data set", color = "blue",
                                   hjust = 1, x = 1, face = "italic", size = 10),
                left = text_grob("Tooth length", color = "green", rot = 90),
                right = "I'm done, thanks :-)!",
                fig.lab = "Figure 1", fig.lab.face = "bold"
)

grid.(
  arrangeGrob(
  combined_plot,
  left = textGrob(
    "Normalized Counts",
    rot = 90,
    vjust = 1,
    gp = gpar(cex = 1.3)
  ),
  bottom = textGrob("Genes", gp = gpar(cex = 1.3))
))

grid.arrange(arrangeGrob(plot, left = y.grob, bottom = x.grob))
# ggarrange(bp.gastroc.up, bp.solues_up, bp.gastroc.down, bp.solues_down, common.legend = T)
```

```{r include=FALSE, eval=FALSE}

plot <- plot_grid(
  bp.gastroc.up + theme(legend.position = "none"),
  bp.soleus.up + theme(axis.text.y = element_blank(), legend.position = "none"),
  bp.gastroc.down + theme(legend.position = "none"),
  bp.soleus.down + theme(axis.text.y = element_blank(), legend.position = "none"),
  align = 'vh',
  vjust = 1,
  scale = 1
)

plot_grid(plot_grid(plot ,
                 plot_grid(NULL, my_legend, ncol=1),
                 rel_widths=c(1, 0.2)), )

my_legend <- get_legend(bp.gastroc.up)


grid.arrange(arrangeGrob(
  plot,
  left = textGrob(
    "Normalized Counts",
    rot = 90,
    vjust = -0.3,
    gp = gpar(cex = 1)
  ),
  bottom = textGrob("Genes", gp = gpar(cex = 1.3))
))
```



## Venn/Euler-Diagram

significant grouped:

```{r vennDiagram}
venn.colors <- c(params$tissue_pal, scales::hue_pal()(4))

# numbers for euler diagram
gene_sets <- c(
  "gastrocnemius" = sign_gene_stats$gastroc - sign_gene_stats$shared_sig_genes,
  "soleus" = sign_gene_stats$soleus - sign_gene_stats$shared_sig_genes,
  "gastrocnemius&soleus" = sign_gene_stats$shared_sig_genes
)

# euler: two tissues
p <- plot(
  euler(gene_sets),
  quantities = list(cex = 1.7),
  legend = list(side = "bottom", cex = 1.7),
  fills = venn.colors
  # main = "significant genes"
)
p
```


```{r vennDiagram_save, include=FALSE}
ggsave(file.path(plot.dir, "02_euler.svg"), p)
ggsave(file.path(plot.dir, "02_euler.png"), p)
```


# Clustering


Replacing the raw counts with normalized counts for better visual representation 
of the performed DEA:

```{r obtaining_normalized_counts}
counts.gastroc <- counts(dds.gastroc, normalized = T)
counts.soleus <- counts(dds.soleus, normalized = T)
```

The genes previously determined significant (`tissue_res.filtered`) in both
tissues will be clustered via hierachical clustering and a PCA will be performed
by taking the union of the genes.

```{r prepare_sign_counts}
# obtaining all DE genes, but avoiding duplicates
sign_genes <-
  unique(c(
    row.names(gastroc_res.filtered),
    row.names(soleus_res.filtered)
  )) 
# making sure, that DE genes occur in both tissues
sign_genes <-
  sign_genes[sign_genes %in% rownames(counts.gastroc) &
               sign_genes %in% rownames(counts.soleus)]

counts_sign <- merge(
  counts.gastroc[sign_genes,],
  counts.soleus[sign_genes,],
  by = 0,
  suffixes = c("_gastrocnemius", "_soleus")
) %>% 
  tibble::column_to_rownames(var="Row.names") %>%
  as.matrix()

```


## Heatmap

A zscore is applied to the counts to visualize the differential expression of the
normalized counts in a heatmap.

```{r heatmap, fig.width=4, fig.height=4, message=FALSE}
zscore <- function(M) {
  s <- apply( M,1,sd )      # standard deviation
  µ <- apply( M, 1, mean )  # mean
  M.z <- (M - µ) / s        # zscore
  return(M.z)
}

# using the function from "code/visualizations_functions.R"
chm <- plot_combined_CHM(zscore(counts_sign), params=params)
```


```{r heatmap_save, include=FALSE}
# saving as .svg
svglite::svglite(file.path(plot.dir, "02_heatmap_sign_genes.svg"), width=6, height=6)
chm
dev.off()

# saving as .png
png(file.path(plot.dir, "02_heatmap_sign_genes.png"), width=6, height=6, units = "in", res = 1200)
chm
dev.off()
```


## pca: significant genes
PCA is applied to the same significantly DE genes (normalized) as used in the heatmap.

```{r pca_sign_counts_normalized, warning=FALSE}
# using the function from "code/visualizations_functions.R"
p <- plot_combined_pca(counts_sign, params=params)
p
```


```{r pca_sign_counts_normalized_save, include=FALSE}
ggsave(file.path(plot.dir, "02_pca_dea_sign.svg"), p)
ggsave(file.path(plot.dir, "02_pca_dea_sign.png"), p)
```



# save R ojects

```{r saving_R_objects}
save(dds.gastroc, dds.soleus, res.gastroc, res.soleus, file = "./data/Robjects/02_DDS.RData")
```