---
title: "03_Annotation"
author: "Nick Diercksen"
date: "Last compiled on `r format(Sys.time())`"
output:
  html_notebook:
    toc: yes
    toc_float: yes
    df_print: paged
  html_document:
    toc: yes
    df_print: paged
editor_options: 
  chunk_output_type: inline
params:
  treemap_pal: !r colorRampPalette(RColorBrewer::brewer.pal(11, "Paired"))
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(dplyr)
library(biomaRt)  # interface to a collection of databases
library(stringr)
library(SummarizedExperiment)
library(ggplot2)
```



```{r plottingSettings, include=FALSE}
ggplot2::theme_set(theme_bw())

plot.dir <- file.path("plots", "03 annotation")
```

# About

This step is a preparation for the following Gene Set Enrichment Analysis  (GSEA with `fgsea`) and GO analysis.

To link the genes with the pathways the Entrez IDs ("entrezgene_id") is needed.
Therefore the `biomaRt` interface is used to access the corresponding DBs.


# read in data

The Ensembl gene IDs can be found in the saved SummarizedExperiment object.
As the genes are the same for gastrocnemius and soleus, only one set needs to be 
loaded. The Ensembl IDs can be used to query the ensembl.org database.

```{r}
se.gastroc <- readRDS("./data/Robjects/01_se.gastroc.rds")
ensembl_ids <- rownames(se.gastroc)
```


# query biomaRt for annotations

For the mouse dataset ("mmusculus_gene_ensembl") the ensembl.org database is queried
for entries of the current Ensembl IDs.
Besides the "entrezgene_id" some further annotation data is requested to compare
already available annotation data.

```{r}
# 
ensembl.mart <- useMart(biomart = "ensembl", dataset = "mmusculus_gene_ensembl", host = 'https://www.ensembl.org')

filterType <- "ensembl_gene_id"
attributeNames <- c('ensembl_gene_id',
                    'entrezgene_id',
                    'external_gene_name',
                    'description',
                    'gene_biotype',
                    'chromosome_name',
                    'start_position',
                    'end_position',
                    'strand')
# run the query
annot <- getBM(attributes=attributeNames,
               filters = filterType,
               values = ensembl_ids,
               mart = ensembl.mart) %>%
  dplyr::rename(GeneID=ensembl_gene_id)


```


# exploring the annotation dataset:

duplicate ids:
```{r}
sum(duplicated(annot$GeneID))
```


## missing genes

```{r}
missingGenes <- ! ensembl_ids %in% annot$GeneID
print(paste0("There are ", sum(missingGenes), " genes not having an entry in the GRCm39 assembly genome"))
ensembl_ids[missingGenes] %>% head(20)
```

```{r treemap_missingGenes}
biotype_colors <- table(rowData(se.gastroc)$gene_biotype, dnn = c("biotype")) %>%
  as.data.frame() %>%
  mutate(
    biotype = stringr::str_replace_all(biotype, "_", " "),
    colors = params$treemap_pal(n())
    ) %>% 
  dplyr::select(-Freq)


genes_without_entrez <- rowData(se.gastroc) %>% as.data.frame() %>% .[missingGenes,]

gene_biotypes <- table(genes_without_entrez$gene_biotype, dnn = c("biotype")) %>%
  as.data.frame() %>%
  mutate(
    biotype = stringr::str_replace_all(biotype, "_", " "),
    ) %>% 
  arrange(desc(Freq)) %>% 
  merge(biotype_colors, by = "biotype") # obtain the same colors

treemap::treemap(gene_biotypes,
                 type = "color",
                 index = "biotype",
                 vSize = "Freq",
                 vColor = "colors",
                 title = "Gene biotypes",
                 fontsize.labels = 20)
```


on the example of gastrocnemius looking at how many of those genes were already filtered out and not important:
```{r pieChart_missingGenes_filtered}
filtered.df <- data.frame(
  group = c("expressed", "not expressed"),
  value = c(sum(genes_without_entrez$filtered), sum(!genes_without_entrez$filtered))
)
ggplot(filtered.df, aes(x="", y=value, fill=group)) +
  geom_bar(stat="identity", width=1) +
  geom_text(aes(label = value),
            position = position_stack(vjust = 0.5), size = 8) +
  coord_polar("y", start=0) + 
  theme_void() +
  theme(legend.text = element_text(size = 17), legend.title = element_blank())
```



## consistency in the metadata (sanity checks)

the metadata from the genes originates from two sources:

* already provided with the available counts
* from biomart (Ensembl)

The question is if both datasets are consistent with the data they provide.
Inconsistencies can point to possible mistakes done when annotating.

So before merging the new annotation data into the `se` to get the "entrezgene_id"
to later do the pathway enrichment.

```{r}
df <- merge(annot, rowData(se.gastroc), by.x = 1, by.y = 0) %>%
  as.data.frame()
```


### biotypes:
```{r}
cols <- c("GeneID","gene_biotype.x", "gene_biotype.y")
dif_biotypes <- !df$gene_biotype.x == df$gene_biotype.y
print(paste0("There are ", sum(dif_biotypes), " genes annotated with different biotypes"))
df[(dif_biotypes),cols] %>% head(20)
```

About a fifth of the genes is annotated with a different biotype. This has to be
looked at more specifically though, since this could still be viable in the 
biological sense.

### gene names
```{r}
cols <- c("GeneID", "external_gene_name", "gene_name")
dif_names <- !df$external_gene_name == df$gene_name
print(paste0("There are ", sum(dif_names), " genes annotated with different gene names"))
df[(dif_names), cols] %>% head(20)
```

Not as many differences as the biotypes, and most of them seem to be synonyms:

* http://www.ensembl.org/Mus_musculus/Gene/Summary?db=core;g=ENSMUSG00000002280;r=17:25992750-26002306
* http://www.ensembl.org/Mus_musculus/Gene/Summary?db=core;g=ENSMUSG00000003527;r=16:17718573-17729212

### chromosome
```{r}
cols <- c("GeneID", "chromosome_name", "gene_chr", "external_gene_name", "gene_name")
df[(!df$chromosome_name == df$gene_chr), cols] %>% head(20)
```

Here we "only" have 2 genes annotated with different chromosomes, but this might
be one of the more severe differences!

**ENSMUSG00000065387**

* "5S_rRNA": is known in mammals (Pan troglodytes[Chimpanzee], Canis familiaris[Dog], Bos Taurus[Cow])
* "n-R5s85": is a gene in "Mus musculus" (What is expected)

| [Source](https://www.genecards.org/cgi-bin/carddisp.pl?gene=RNA5SP273)

=> Could this mean, that external_gene_name is wrongly matched to other species than mouse?


**ENSMUSG00000076377**

* "Gm22370": is assigned to Mouse as "ENSMUST00000102422.1" (ensembl transcript id) [Source](http://rnact.crg.eu/rnas?page=228)




### gene positions
#### start
```{r}
dif_start <- df$start_position - df$gene_start
head(dif_start, 20)
summary(dif_start)
```

total genes with the same start position
```{r}
sum(dif_start == 0)
```

```{r}
hist(abs(dif_start), breaks = 100)
```

These differ quite substantially. Either these are different metrics or really
relating to a different species.

#### end
```{r}
dif_end <- df$end_position - df$gene_end
head(dif_end, 20)
summary(dif_end)
```


total genes with the same end position

```{r}
sum(dif_end == 0)
```

```{r}
hist(abs(dif_end), breaks = 100)
```

Similar to the start.

## Discussion
On the first glance, there are several inconsistencies with the two annotation
sources. Whether and where a wrong data source was chosen has to be examined.

Gene length was not compared, since no field was available and subtracting the
start from the end position would not make sense (introns would probably 
get counted as well).


# reviewing joined annotations

As some genes share the same Entrez IDs, several of those are displayed in the
following.

## duplicated Entrez IDs

```{r}
# getting all genes which have entrez gene id assigned
rdata.entrez <- as.data.frame(rowData(se.gastroc)) %>%
    tibble::rownames_to_column("GeneID") %>%
    left_join(annot, by=c("GeneID")) %>%
    filter(!is.na(entrezgene_id))

print(paste0(nrow(rdata.entrez), "/", nrow(se.gastroc), " genes were annotated with entrez gene ids"))

# getting duplicate
dups.entrez <- rdata.entrez$entrezgene_id %>%
  base::duplicated() %>%
  which() %>%
  rdata.entrez$entrezgene_id[.]

subset(rdata.entrez, entrezgene_id %in% dups.entrez) %>%
  dplyr::select(GeneID, entrezgene_id, external_gene_name) %>%
  arrange(entrezgene_id, ) %>% 
  head()
```

## Genes missing Entrez IDs

Biotypes of the genes, which could not be assigned a Entrez ID and thus will be 
omitted in the subsequent enrichment analysis steps.

```{r treemap_missingEntrezIds}
gene_biotypes <- table(rdata.entrez$gene_biotype.x, dnn = c("biotype")) %>%
  as.data.frame() %>%
  mutate(
    biotype = stringr::str_replace_all(biotype, "_", " "),
    ) %>% 
  arrange(desc(Freq)) %>% 
  merge(biotype_colors, by = "biotype") # obtain the same colors

treemap::treemap(gene_biotypes,
                 type = "color",
                 index = "biotype",
                 vSize = "Freq",
                 vColor = "colors",
                 title = "Gene biotypes", 
                 fontsize.labels = 20)
```


on the example of gastrocnemius looking at how many of those genes were already filtered out and not important:

```{r pieChart_missingEntrez_filtered}
filtered.df <- data.frame(
  group = c("expressed", "not expressed"),
  value = c(sum(rdata.entrez$filtered), sum(!rdata.entrez$filtered))
)
ggplot(filtered.df, aes(x="", y=value, fill=group)) +
  geom_bar(stat="identity", width=1) +
  geom_text(aes(label = value),
            position = position_stack(vjust = 0.5), size = 8) +
  coord_polar("y", start=0) + 
  theme_void() +
  theme(legend.text = element_text(size = 17), legend.title = element_blank())
```


## duplicate ENSMUS

ENSMUS identifier are mostly transcripts of genes. Thus there are some genes
mapping to multiple ENSMUS ids.
! this might have been induced by the joining of the downloaded annotations, which
already had duplicate entrez IDS.


```{r}
# genes in GRCm39 that have 
dups.GeneID <- annot$GeneID %>%
  base::duplicated() %>%
  annot$GeneID[.] %>% 
  unique()

subset(annot, GeneID %in% dups.GeneID) %>%
  dplyr::select(GeneID, entrezgene_id, external_gene_name) %>%
  arrange(GeneID, )

dups <- annot %>%
  dplyr::select("GeneID", "external_gene_name") %>%
  duplicated()

# this implies some genes can be annotated with several pathways and rowData can't just be extended
# does something

annot_withDups <- annot
annot <- annot_withDups[!dups,]
```




# save Data

```{r}
# TODO: merge `annot` into se metadata, rowData is not compatible?
save(annot, file = "./data/Robjects/03_annot.RData")
```

this is currently used again in `04_fgsea`.