---
title: "03_DESeq"
author: "Nick Diercksen"
output:
  html_notebook:
    code_folding: hide
    toc: yes
    toc_float: yes
    df_print: paged
  html_document:
    toc: yes
    df_print: paged
editor_options:
  chunk_output_type: inline
params:
  FCutoff:
    label: cutoff for the log2FoldChange
    value: 1
  pCutoff:
    label: "cutoff for the adjusted p-Value"
    value: 0.01
  condition_pal: !r c("WT"= "#00C3C6","KO"= "#FF6C67")
  tissue_pal: !r c("gastroc" = "orange", "soleus"="purple")
  regulated_pal: !r list("upregulated" = 'royalblue', "downregulated" = 'red', "insignificant" = 'gray')
  heatmap_pal: !r rev(RColorBrewer::brewer.pal(11, "RdYlBu"))
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(dplyr)
library(ggplot2)
library(ggfortify) # autoplot (PCA)
library(DESeq2)
# TODO: maybe just remove SE altogether again?! seems to just bloat everything
library(SummarizedExperiment)

library(ComplexHeatmap)
library(EnhancedVolcano)
library(eulerr)
library(RColorBrewer)
```



# loading data

For the DESeq2 analysis, the raw, prefiltered counts will be used.

```{r loadRDS}
se.gastroc <- readRDS("./data/Robjects/01_se.gastroc.rds")
se.soleus <- readRDS("./data/Robjects/01_se.soleus.rds")

counts.gastroc <- se.gastroc[rowData(se.gastroc)$filtered,] %>% assay()
counts.soleus <- se.soleus[rowData(se.soleus)$filtered,] %>% assay()
metadata <- colData(se.gastroc)
```

# preparing DDS objects

```{r prepDDS, message=FALSE, warning=FALSE}
# trying to reorder the metadata, so that the design is correct (so far no luck)
# metadata_2 <- as.data.frame(metadata)
# metadata_2$genotype <- factor(metadata_2$genotype, levels = c("KO", "WT"))

# metadata_2 <- rbind(metadata[7:12,,drop=FALSE], metadata[1:6,,drop=FALSE])

metadata$genotype <- as.factor(metadata$genotype) # for DESeq



createDDSObject <- function(counts, metadata) {
  # select sample columns
  reorder_index <- match(rownames(metadata), colnames(counts))
  counts <- counts[,reorder_index] %>% 
    apply(c(1,2), as.integer) # DESeq need normal counts (integer) as input
  
  # Check metadata consistency
  all(rownames(metadata) %in% colnames(counts)) %>%
    assertthat::assert_that(., msg = "metadata and count table do not match")
  
  ## DESeq2 object
  dds <- DESeqDataSetFromMatrix(countData = counts,
                                colData = metadata,
                                design = ~ genotype)
  
  return( DESeq(dds) )
}

# creating DESeq Objects
dds.gastroc <- createDDSObject(counts.gastroc, metadata)
dds.soleus <- createDDSObject(counts.soleus, metadata)
```



# Results

```{r results}
res.gastroc <- results(dds.gastroc, alpha = params$pCutoff, contrast = c("genotype", "KO", "WT"))
res.soleus <- results(dds.soleus, alpha = params$pCutoff, contrast = c("genotype", "KO", "WT"))
```



```{r lfc}
# lfcShrink is not applied by default

# library(ashr) # using instead of "apeglm", because https://www.bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#extended-section-on-shrinkage-estimators

# resultsNames(dds.gastroc)
resLFC.gastroc <- lfcShrink(dds.gastroc, contrast = c("genotype", "KO", "WT"), type = "ashr")
resLFC.soleus <- lfcShrink(dds.soleus, contrast = c("genotype", "KO", "WT"), type = "ashr")
```




## count outlier detection with cooks distance

Filtering outliers using the cooks distance is done automatically and p-values are set to `NA`.
Changing (lowering) the `minReplicatesForReplace` will replace these flagged values.

DESeq2 flags these outliers and removes the corresponding genes completely, if below 7 samples (above would replace the values)


```{r}
# assuming all NA p-values are the ones filtered out using the cooks distance

print_outliers <- function(res, se){
  # all genes with a `NA` value are exactly those 5 which were also detected by the cooks distance:
  outlier_ids <- res %>% 
    as.data.frame() %>% 
    filter(is.na(pvalue)) %>% 
    rownames()
  
  # a different approach would be finding the genes which have a cooks distance
  # above the cutoff:
  # cooksCutoff <- qf(.99, 2,10)
  # outlier_ids <- which(mcols(dds)$maxCooks > cooksCutoff) %>% names()
  
  # getting counts and metadata
  outlier_metadata.df <- rowData(se) %>%
    as.data.frame() %>%
    select(gene_name, gene_biotype)
  
  counts.df <- assay(se)
  
  # merging df
  outliers.df <-
    merge(outlier_metadata.df, counts.df[outlier_ids,], by = 0)
  return(outliers.df)
}

outliers_gastroc.df <- print_outliers(res.gastroc, se.gastroc) %>% tibble::column_to_rownames(var = "Row.names")
write.csv(outliers_gastroc.df, file = "./code/analysis/03_outliers_gastroc.csv")
outliers_gastroc.df %>% knitr::kable()
print_outliers(res.soleus, se.soleus) %>% knitr::kable()
```


## size factors
```{r size_factors}
sizeFactors(dds.gastroc)
sizeFactors(dds.soleus)
```



## Dispersion Estimates

### gastroc

```{r dispEst.gastroc}
plotDispEsts(dds.gastroc)
```

### soleus

```{r dispEst.soleus}
plotDispEsts(dds.soleus)
```


## MA-Plots

### gastroc

```{r ma.gastroc, message=FALSE, warning=FALSE}
plotMA(res.gastroc, ylim = c(-6,6))
```

```{r maLFC.gastroc, message=FALSE, warning=FALSE, include=FALSE, eval=FALSE}
# log fold change shrinked:
plotMA(resLFC.gastroc)
```

### soleus

```{r ma.soleus, message=FALSE, warning=FALSE}
plotMA(res.soleus, ylim = c(-6,6))
```

```{r maLFC.soleus, message=FALSE, warning=FALSE, include=FALSE, eval=FALSE}
# log fold change shrinked:
plotMA(resLFC.soleus)
```


## top significant diff. genes

```{r topDiffGenes.df}
topGenes.gastroc <- as.data.frame(res.gastroc) %>%
  tibble::rownames_to_column("GeneID") %>%
  top_n(100, wt = -padj) %>%
  arrange(padj)

topGenes.soleus <- as.data.frame(res.soleus) %>%
  tibble::rownames_to_column("GeneID") %>%
  top_n(100, wt = -padj) %>%
  arrange(padj)

knitr::kable(head(topGenes.gastroc), caption = "gastroc")
knitr::kable(head(topGenes.soleus), caption = "soleus")
```

## p-values

```{r pValuesHist}
# TODO: use ggplot for unified look of plots
hist(res.gastroc$pvalue, main = "gastroc")
hist(res.soleus$pvalue, main = "soleus")
```


## Volcano Plot
```{r volcanoPlot, warning=FALSE, fig.width = 9, fig.height = 9, fig.retina = 2, dpi = 100}
volcanoPlot <- function(result, se, pCutoff = 0.01, FCutoff = 1, tissue = character()) {
  gene_names <-
    rowData(se)[rownames(result), c("gene_name"), drop = F]
  results.df <- result %>%
    as.data.frame() %>%
    dplyr::arrange(padj)
  
  # top 10 gene labels for respectively up and down regulation
  labs.up <- results.df[results.df$log2FoldChange > FCutoff, ] %>%
    rownames() %>% .[1:10] %>% gene_names[., c("gene_name")]
  labs.down <- results.df[results.df$log2FoldChange < -FCutoff, ] %>%
    rownames() %>% .[1:10] %>% gene_names[., c("gene_name")]
  selectLab <- c(labs.up, labs.down, "Nfe2l1") %>% unique() # always including "Nfe2l1"
  
  # custom colors:
  keyvals <- ifelse(
    result$log2FoldChange > FCutoff &
      result$padj < pCutoff,
    params$regulated_pal$upregulated,
    ifelse(
      result$log2FoldChange < -FCutoff &
        result$padj < pCutoff,
      params$regulated_pal$downregulated,
      params$regulated_pal$insignificant
    )
  )

  keyvals[is.na(keyvals)] <- params$regulated_pal$insignificant
  names(keyvals)[keyvals == params$regulated_pal$upregulated] <- 'up regulated'
  names(keyvals)[keyvals == params$regulated_pal$insignificant] <- 'nonsignificant'
  names(keyvals)[keyvals == params$regulated_pal$downregulated] <- 'down regulated'
  
  vlc.plt <- EnhancedVolcano(
    result,
    x = 'log2FoldChange',
    y = 'padj',
    title = 'KO vs WT: Nfe2l1 knockout',
    subtitle = ifelse(isEmpty(tissue), "", paste0('tissue: ', tissue)),
    caption = "",
    ylab = expression(paste(-Log[10], p[adj])),
    pCutoff = pCutoff,
    FCcutoff = FCutoff,
    legendPosition = 'right',
    pointSize = 2,
    colCustom = keyvals,
    lab = gene_names$gene_name,
    selectLab = selectLab,
    labSize = 3,
    boxedLabels = TRUE,
    drawConnectors = TRUE,
    max.overlaps = Inf
  )
  
  return(vlc.plt)
}
```


```{r volcano, warning=FALSE, fig.width = 9, fig.height = 9, fig.retina = 2, dpi = 100}
p <- volcanoPlot(res.gastroc, se.gastroc, pCutoff = params$pCutoff, FCutoff = params$FCutoff, tissue = "gastrocnemius")
ggsave(filename = "plots/02_volcano_gastroc.svg", p)
ggsave(filename = "plots/02_volcano_gastroc.png", p)
p
p <- volcanoPlot(res.soleus, se.soleus, pCutoff = params$pCutoff, FCutoff = params$FCutoff, tissue = "soleus")
ggsave(filename = "plots/02_volcano_soleus.svg", p)
ggsave(filename = "plots/02_volcano_soleus.png", p)
p
```



```{r volcanoLFC, fig.width = 9, fig.height = 9, fig.retina = 2, dpi = 100}

p <- volcanoPlot(resLFC.gastroc, se.gastroc, pCutoff = params$pCutoff, FCutoff = params$FCutoff, tissue = "gastrocnemius")
ggsave(filename = "plots/02_volcano_gastroc_LFC.svg", p)
ggsave(filename = "plots/02_volcano_gastroc_LFC.png", p)
p
p <- volcanoPlot(resLFC.soleus, se.soleus, pCutoff = params$pCutoff, FCutoff = params$FCutoff, tissue = "soleus")
ggsave(filename = "plots/02_volcano_soleus_LFC.svg", p)
ggsave(filename = "plots/02_volcano_soleus_LFC.png", p)
p
```



## scatter-plot: most differential genes, both tissues

using the Wald-test `stat` from the DESeq2 result and filtering on the set
`FCutoff=``r params$FCutoff` and `pCutoff=``r params$pCutoff` yields the following plot:
```{r scatterDEA, warning=FALSE, fig.width = 9, fig.height = 9, fig.retina = 2, dpi = 100}
apply_cutoffs <- function(deseq.result, colname="stat", FCutoff, pCutoff) {
  res.filtered <- deseq.result %>%
    data.frame() %>%
    filter(padj < pCutoff,
           log2FoldChange > FCutoff | log2FoldChange < -FCutoff) %>%
    dplyr::rename(!!colname := stat) %>%
    dplyr::select(!!colname)
  return(res.filtered)
}

gastroc_res.filtered <- apply_cutoffs(res.gastroc, colname="gastroc", params$FCutoff, params$pCutoff)
soleus_res.filtered  <- apply_cutoffs(res.soleus,  colname="soleus",  params$FCutoff, params$pCutoff)

gene_names <- rowData(se.gastroc) %>% as.data.frame() %>% 
  dplyr::select(gene_name)

# combining Wald-Test data from both tissues and ordering in quadrants
res.combined <- merge(gastroc_res.filtered,
                      soleus_res.filtered,
                      by = 0) %>%
  mutate(diff.exp = case_when(
    gastroc < 0 & soleus < 0 ~ "both down",
    gastroc > 0 & soleus > 0 ~ "both up",
    gastroc < 0 & soleus > 0 ~ "gastrocnemius down,\n soleus up",
    gastroc > 0 & soleus < 0 ~ "gastrocnemus up,\n soleus down",
    TRUE ~ "different"
  )) %>% 
  merge(gene_names, by.x="Row.names", by.y=0)
 
# removing all gene_names except the top_n_genes (sum of absolute Wald-test numbers)
top_n_genes <- 10
top_labels <- res.combined %>%
  group_by(diff.exp) %>% 
  arrange(desc(abs(gastroc) + abs(soleus))) %>%
  filter(row_number() %in% c(1:top_n_genes)) %>% 
  ungroup() %>% 
  .$gene_name

res.combined <- res.combined %>% 
  mutate(gene_name = ifelse(gene_name %in% top_labels, gene_name, ""))

# final plot
p <- ggplot(res.combined, aes(x = gastroc, y = soleus, label = gene_name)) +
  geom_vline(xintercept = 0) + 
  geom_hline(yintercept = 0) + 
  geom_point(aes(color = diff.exp)) +
  # scale_color_manual(values = c("red", "chartreuse1", "bisque", "royalblue")) +
  labs(x = "gastrocnemius", y = "soleus", color = "significantly\ndifferentially\nexpressed") +
  ggrepel::geom_label_repel(max.overlaps = 15) + 
  ggtitle(label = "DEA: t-statistics (Wald test)") +
  theme_bw() +
  theme(
    axis.title = element_text(size = 13),
    axis.text = element_text(size = 13),
    title = element_text(size = 13),
    legend.text = element_text(size = 10)
  )

ggsave(filename = "plots/02_dea_scatter.svg", p)
ggsave(filename = "plots/02_dea_scatter.png", p)
p
```


### barplot

```{r dea_scatter_barplot}
p <- ggplot(res.combined, aes(x = diff.exp)) +
  geom_bar(aes(fill = diff.exp)) +
  theme_bw()
ggsave(filename = "plots/02_dea_scatter_barplot.svg", p)
ggsave(filename = "plots/02_dea_scatter_barplot.png", p)
p
```

```{r include=FALSE}
# obtain gene counts for the respective groups for visualizing
sign_gene_stats <- list(
  "all genes" = nrow(gene_names),
  "gastrocnemius" = nrow(gastroc_res.filtered),
  "soleus" = nrow(soleus_res.filtered),
  "shared_sig_genes" = nrow(res.combined),
  "both up" = sum(res.combined$diff.exp == "both up"),
  "both down" = sum(res.combined$diff.exp == "both down"),
  "ga up, sol down" = sum(res.combined$diff.exp == "ga up, sol down"),
  "ga down, sol up" = sum(res.combined$diff.exp == "ga down, sol up")
)
```


## boxplots: significant genes in both tissues


```{r boxplotSignGenes, message=FALSE, warning=FALSE}

boxplot.genes <- function(counts.df, title ="", FCutoff = 1) {
  
  ggplot(counts.df,
         aes(
           # x = as.factor(gene_name),
           x = reorder(gene_name,ensembl),
           y = normalized_counts,
           fill = genotype
         )) +
    geom_dotplot(
      binaxis = 'y',
      stackdir = 'center',
      dotsize = 0.3,
      position = position_dodge(0.8),
      fill = "black"
    ) +
    geom_boxplot(outlier.size = 0.3) +
    scale_y_log10() +
    xlab("Genes") +
    ylab("Normalized Counts") +
    scale_fill_manual(values = condition_pal) +
    ggtitle(title) +
    theme_bw() +
    theme(plot.title = element_text(hjust = 0.5)) +
    theme(axis.text.x = element_text(
      angle = 45,
      hjust = 1,
      face = setBold(counts.df$gene_name, c("Nfe2l1"))
    ))
}

prepare_counts <- function(gene_ids, dds, se, reorder_genes = F) {
  counts.top <- counts(dds, normalized = T)[gene_ids, ]
  metadata <- colData(se) %>% as.data.frame()
  gene_names <-
    rowData(se)[, c("gene_name"), drop = F] %>% as.data.frame()
  
  # the prepared counts df for the plot
  counts.plt <-
    data.frame(counts.top) %>%
    tibble::rownames_to_column(var = "ensembl") %>%
    tidyr::gather(key = "samplename",
                  value = "normalized_counts", 2:13) %>%
    merge(metadata, by.x = "samplename", by.y = 0) %>%
    merge(gene_names, by.x = "ensembl", by.y = 0)  %>%
    mutate(genotype = factor(genotype)) %>%
    mutate(genotype = relevel(genotype, "WT")) # "WT" needs to be displayed before "KO"
    
    if (reorder_genes) {
      counts.plt <- counts.plt %>%
        mutate(
          gene_name = forcats::fct_reorder(gene_name, normalized_counts, .desc = T),
          ensembl = forcats::fct_reorder(ensembl, normalized_counts, .desc = T)
        )
    } else {
      idx = match(counts.plt$ensembl, gene_ids)
      counts.plt <- counts.plt[order(idx),] %>%
        mutate(gene_name = factor(gene_name))# %>%
        # arrange(gene_name)
      counts.plt$ensembl <- factor(counts.plt$ensembl, levels = gene_ids,ordered = TRUE)
      counts.plt$gene_name <- factor(counts.plt$gene_name)

      # counts.plt$gene_name <- factor(counts.plt$gene_name, levels = gene_ids)
      # counts.plt <- counts.plt %>% 
      #   mutate(gene_name = factor(gene_name, levels = gene_ids)) %>% 
      #   arrange()
    }
  return(counts.plt)
}


# helper function to set the font of a label on the axis to bold
setBold <- function(src, special_labs) {
  # source: https://stackoverflow.com/questions/39694490/highlighting-individual-axis-labels-in-bold-using-ggplot2
  if (!is.factor(src))
    src <- factor(src)
  src_levels <- base::levels(src)
  brave <- special_labs %in% src_levels
  b_vec <- rep("plain", length(src_levels))
  if (all(brave)) {
    b_pos <- purrr::map_int(special_labs, ~ which(. == src_levels))
    b_vec[b_pos] <- "bold"
    b_vec
  } else {
    message("setBold: no matching element found")
  }
  return(b_vec)
}

# actual plot call
for (group in unique(res.combined$diff.exp)) {
  gene_ids <- filter(res.combined, diff.exp == group)$`Row.names`
  prep_counts.gastroc <- prepare_counts(gene_ids, dds.gastroc, se.gastroc, reorder_genes = T)
  prep_counts.soleus <-
    prepare_counts(levels(prep_counts.gastroc$ensembl), dds.soleus, se.soleus, reorder_genes = T)
  gene_ids <- levels(prep_counts.gastroc$ensembl)
  
  p.g <- boxplot.genes(
    counts.df = prep_counts.gastroc,
    title = paste0("Sign. regulated genes: ", group, "\n tissue: ", "gastrocnemius")
  ) #%>% print()
  
  p.s <- boxplot.genes(
    counts.df = prep_counts.soleus,
    title = paste0("\n tissue: ", "soleus")
  ) #%>% print()
  
  ggpubr::ggarrange(p.g, p.s, ncol = 1, nrow = 2)
}
```


## boxplots: top 20 significant genes

```{r topDiffGenes}
setBold <- function(src, special_labs) {
  # source: https://stackoverflow.com/questions/39694490/highlighting-individual-axis-labels-in-bold-using-ggplot2
  if (!is.factor(src))
    src <- factor(src)
  src_levels <- base::levels(src)
  brave <- special_labs %in% src_levels
  b_vec <- rep("plain", length(src_levels))
  if (all(brave)) {
    b_pos <- purrr::map_int(special_labs, ~ which(. == src_levels))
    b_vec[b_pos] <- "bold"
    b_vec
  } else {
    message("setBold: no matching element found")
  }
  return(b_vec)
}

getTopExpressedEnsemblNames <- function(result,
                                        FCutoff,
                                        n,
                                        up = T) {
  .filter <- ifelse(up, `>`, `<`)
  
  subset(result, .filter(log2FoldChange, FCutoff)) %>%
    data.frame() %>%
    filter(baseMean > 100) %>%
    arrange(padj) %>%
    .[1:n, ] %>%
    rownames()
}

boxplot.top <- function(result, dds, se, upregulated=TRUE, n = 20, FCutoff = 1, tissue ="") {
  # getting the top n regulated genes
  names.top <- getTopExpressedEnsemblNames(result, FCutoff=FCutoff, n=n, up = upregulated)
  
  counts.top <- counts(dds, normalized=T)[names.top, ]
  metadata <- colData(se) %>% as.data.frame()
  gene_names <- rowData(se)[, c("gene_name"), drop = F] %>% as.data.frame()
  
  counts.plt <-
    data.frame(counts.top) %>%
    tibble::rownames_to_column(var = "ensembl") %>%
    tidyr::gather(key = "samplename",
           value = "normalized_counts", 2:13) %>%
    merge(metadata, by.x="samplename", by.y=0) %>%
    merge(gene_names, by.x="ensembl", by.y=0)  %>%
    mutate(genotype = factor(genotype)) %>%
    mutate(genotype = relevel(genotype, "WT"))  %>% # "WT" needs to be displayed before "KO"
    mutate(gene_name = forcats::fct_reorder(gene_name, normalized_counts, .desc = T))
  
  direction <- ifelse(upregulated, "up", "down")
  
  ggplot(counts.plt,
         aes(
           x = as.factor(gene_name),
           y = normalized_counts,
           fill = genotype
         )) +
    geom_dotplot(
      binaxis = 'y',
      stackdir = 'center',
      dotsize = 0.3,
      position = position_dodge(0.8),
      fill = "black"
    ) +
    geom_boxplot(outlier.size = 0.3) +
    scale_y_log10() +
    xlab("Genes") +
    ylab("Normalized Counts") +
    scale_fill_manual(values = condition_pal) +
    ggtitle(paste0("Top ", n, " ", direction, "-regulated Genes\n tissue: ", tissue)) +
    theme_bw() +
    theme(plot.title = element_text(hjust = 0.5)) +
    theme(axis.text.x = element_text(
      angle = 45,
      hjust = 1,
      face = setBold(counts.plt$gene_name, c("Nfe2l1"))
    ))
}
```

### gastroc

```{r topDiffBox.gastroc, warning=FALSE, message=FALSE}
boxplot.top(res.gastroc, dds.gastroc, se.gastroc, upregulated = T, tissue = "gastrocnemius")
boxplot.top(res.gastroc, dds.gastroc, se.gastroc, upregulated = F, tissue = "gastrocnemius")
```

### soleus

```{r topDiffBox.soleus, warning=FALSE, message=FALSE}
boxplot.top(res.soleus, dds.soleus, se.soleus, upregulated = T, tissue = "soleus")
boxplot.top(res.soleus, dds.soleus, se.soleus, upregulated = F, tissue = "soleus")
```


## Venn/Euler-Diagram

Here are some Venn or Euler (proportional Venn) diagrams to choose from.

Note: using Eulerr diagrams, yields inconsistent plots where the circle sets 
will be placed at different absolute positions each time the plot function is
called.
=> Thus the plot might need to be called multiple times until the wanted 
constellation is obtained.

significant overview with all genes:
```{r}
venn.colors <- c(genes = "white", params$tissue_pal)

gene_sets <- c(
  "all genes" = sign_gene_stats$`all genes` - sign_gene_stats$shared_sig_genes,
  "all genes&gastrocnemius" = sign_gene_stats$gastroc - sign_gene_stats$shared_sig_genes,
  "all genes&soleus" = sign_gene_stats$soleus - sign_gene_stats$shared_sig_genes,
  "all genes&gastrocnemius&soleus" = sign_gene_stats$shared_sig_genes
)

# 1st option: euler
plot(euler(gene_sets),
     legend = list(side = "right") ,
     main = "significant genes per tissue",
     fills = venn.colors)

# 2nd option: venn
plot(venn(gene_sets),
     main = "significant genes per tissue",
     fills = venn.colors)
```



significant grouped:

```{r vennDiagram}
col_palette <- RColorBrewer::brewer.pal(8, "Dark2")
# venn.colors <- c(gastroc = "#FFB08E", soleus = "#FF6638", col_palette[1:4])
venn.colors <- c(params$tissue_pal, scales::hue_pal()(4))

# sets for the venn/euler diagram (exclusive counts)
gene_sets <- c(
  "gastrocnemius" = sign_gene_stats$gastroc - sign_gene_stats$shared_sig_genes,
  "soleus" = sign_gene_stats$soleus - sign_gene_stats$shared_sig_genes,
  "gastrocnemius&soleus" = sign_gene_stats$shared_sig_genes,
  "gastrocnemius&soleus&both down" = sign_gene_stats$`both down`,
  "gastrocnemius&soleus&both up" = sign_gene_stats$`both up`,
  "gastrocnemius&soleus&ga up, sol down" = sign_gene_stats$`ga up, sol down`,
  "gastrocnemius&soleus&ga down, sol up" = sign_gene_stats$`ga down, sol up`
)
# 'gastroc&soleus': if this line is omitted, then all intersects will be denoted 
# with 0 (which was the goal to omit)

# 1st option: eulerr
plot(
  euler(gene_sets),
  quantities = T,
  legend = list(side = "right"),
  fills = venn.colors,
  main = "significant genes"
)


# only the two tissues
gene_sets <- c(
  "gastrocnemius" = sign_gene_stats$gastroc - sign_gene_stats$shared_sig_genes,
  "soleus" = sign_gene_stats$soleus - sign_gene_stats$shared_sig_genes,
  "gastrocnemius&soleus" = sign_gene_stats$shared_sig_genes
)


# 2nd option: euler two tissues
p <- plot(
  euler(gene_sets),
  quantities = T,
  legend = list(side = "right"),
  fills = venn.colors,
  main = "significant genes"
)
ggsave("./plots/03_euler.svg", p)
ggsave("./plots/03_euler.png", p)
p

# 2nd option: venn two tissues
library(eulerr)
p <- plot(
  eulerr::venn(gene_sets),
  fills = params$tissue_pal,
  main = "significant genes"
)
ggsave("./plots/03_venn.svg", p)
ggsave("./plots/03_venn.png", p)
p
```


# Heatmap

Replacing the raw counts with normalized counts for better visual representation 
of the performed DEA:
```{r obtaining_normalized_counts}
counts.gastroc <- counts(dds.gastroc, normalized = T)
counts.soleus <- counts(dds.soleus, normalized = T)
```

```{r prepare_sign_counts}
zscore <- function(M) {
  s <- apply( M,1,sd )      # standard deviation
  µ <- apply( M, 1, mean )  # mean
  M.z <- (M - µ) / s        # zscore
  return(M.z)
}

# obtaining all DE genes, but avoiding duplicates
sign_genes <-
  unique(c(
    row.names(gastroc_res.filtered),
    row.names(soleus_res.filtered)
  )) 
# making sure, that DE genes occur in both tissues
sign_genes <-
  sign_genes[sign_genes %in% rownames(counts.gastroc) &
               sign_genes %in% rownames(counts.soleus)]

counts_sign <- merge(
  counts.gastroc[sign_genes,],
  counts.soleus[sign_genes,],
  by = 0,
  suffixes = c("_gastrocnemius", "_soleus")
) %>% 
  tibble::column_to_rownames(var="Row.names") %>%
  as.matrix()

```


```{r heatmap, fig.width=4, fig.height=4, message=FALSE}
plotCHM <- function(counts.df) {
  # http://colorbrewer2.org/#type=sequential&scheme=BuGn&n=3
  morecols <- colorRampPalette(params$heatmap_pal)
  
  tissue_vec <- c(rep("gastroc", 12), rep("soleus", 12))
  condition_vec <- c(rep(c(rep("WT", 6), rep("KO", 6)),2))
  top_annot <-
    HeatmapAnnotation(
      tissue = tissue_vec,
      condition = condition_vec,
      col = list(tissue = params$tissue_pal, condition = params$condition_pal),
      gp = gpar(col = "darkgray"),
      show_legend = FALSE
    )
  
  ## sample numbers
  sample_nrs <- colnames(counts.df) %>% 
    stringr::str_extract("(?<=_)\\d{1,2}(?=_)")
  colnames(counts.df) = sample_nrs
  
  chm <- Heatmap(
    counts.df,
    row_title = "differentially expressed genes",
    name = "zscore",
    show_row_names = FALSE,
    show_column_names = TRUE,
    column_title = "samples",
    col = morecols(50),
    column_title_side = "bottom",
    top_annotation = top_annot,
    column_names_gp = gpar(col = rep(unname(params$tissue_pal), each=12))
  )
  
  # creating custom annotation legend (to obtain the gray border)
  condition_legend = Legend(
    labels = c("WT", "KO"),
    legend_gp = gpar(fill = params$condition_pal),
    border = "darkgray",
    title = "condition"
  )
  tissue_legend = Legend(
    labels = c("gastrocnemius", "soleus"),
    legend_gp = gpar(fill = unname(params$tissue_pal)),
    border = "darkgray",
    title = "tissue"
  )
  legend_list <- list(condition_legend, tissue_legend)
  
  draw(chm, annotation_legend_list = legend_list)
}

chm <- plotCHM(zscore(counts_sign))

svglite::svglite("./plots/03_heatmap_sign_genes.svg", width=6, height=6)
chm
dev.off()
png("./plots/03_heatmap_sign_genes.png", width=6, height=6, units = "in", res = 1200)
chm
dev.off()
```


### pca: significant genes
PCA is applied to the same significantly DE genes (normalized) as used in the heatmap.

```{r pca_sign_counts_normalized}
pca <- prcomp(t(counts_sign), scale. = T)

pca.data <- data.frame(Sample = rownames(pca$x),
                     X = pca$x[, 1],
                     Y = pca$x[, 2]) %>%
mutate(condition = substr(Sample, 1, 2),
       tissue = stringr::str_extract(Sample,"[:alpha:]+$"),
       sample_nr = stringr::str_extract(Sample, "(?<=_)\\d{1,2}(?=_)"))

p <-
autoplot(
  pca,
  data = pca.data,
  colour = 'tissue',
  shape = 'condition',
  label.label = "sample_nr",
  label = TRUE,
  label.show.legend = FALSE,
  label.repel = TRUE
) +
  # ggtitle("PCA significantly DE genes") +
  scale_color_manual(values = unname(params$tissue_pal)) +
  theme_bw()
ggsave(filename = "plots/02_pca_dea_sign.svg", p)
ggsave(filename = "plots/02_pca_dea_sign.png", p)
p
```


# save R ojects

```{r saving_R_objects}
save(dds.gastroc, dds.soleus, res.gastroc, res.soleus, file = "./data/Robjects/03_DDS.RData")
```